// baqend/commonjs/lib/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logout = exports.ValidateUser = exports.Me = exports.Register = exports.Login = exports.UpdateField = exports.UpdatePartially = exports.CommitTransaction = exports.NewTransaction = exports.GetQueryParameters = exports.RunQuery = exports.GetQueryCode = exports.ListThisQueryResources = exports.CreateQuery = exports.ListQueryResources = exports.AdhocCountQueryPOST = exports.AdhocCountQuery = exports.AdhocQueryPOST = exports.AdhocQuery = exports.DeleteSchema = exports.ReplaceSchema = exports.UpdateSchema = exports.GetSchema = exports.ReplaceAllSchemas = exports.UpdateAllSchemas = exports.GetAllSchemas = exports.DeleteObject = exports.ReplaceObject = exports.GetObject = exports.CreateObject = exports.TruncateBucket = exports.ImportBucket = exports.ExportBucket = exports.GetBucketIds = exports.GetBucketNames = exports.UnbanIp = exports.Unban = exports.Banned = exports.BannedIp = exports.EventsUrl = exports.Status = exports.Connect = exports.UpdateOrestesConfig = exports.GetOrestesConfig = exports.DeleteBloomFilter = exports.GetBloomFilterExpirations = exports.GetBloomFilter = exports.Specification = exports.ApiVersion = exports.ListAllResources = void 0;
exports.Beacon = exports.SW = exports.Install = exports.UploadAPNSCertificate = exports.GCMAKey = exports.VAPIDPublicKey = exports.VAPIDKeys = exports.DeviceRegistered = exports.DevicePush = exports.DeviceRegister = exports.DropAllIndexes = exports.CreateDropIndex = exports.ListIndexes = exports.CleanUpStorage = exports.CleanUpAssets = exports.GetAllRevalidationStatus = exports.CancelRevalidation = exports.GetRevalidationStatus = exports.RevalidateAssets = exports.DownloadAsset = exports.CreateManifest = exports.DeleteFile = exports.UpdateFileMetadata = exports.GetFileMetadata = exports.UploadFile = exports.DownloadFile = exports.CreateFile = exports.DeleteFileBucket = exports.SetFileBucketMetadata = exports.GetFileBucketMetadata = exports.UploadPatchArchive = exports.DownloadArchive = exports.ListBuckets = exports.ListFiles = exports.GetAllModules = exports.GetBaqendModule = exports.PostBaqendModule = exports.DeleteBaqendCode = exports.SetBaqendCode = exports.GetBaqendCode = exports.RevokeUserToken = exports.UserToken = exports.OAuth1 = exports.OAuth2 = exports.VerifyUsername = exports.ChangeUsername = exports.Verify = exports.ResetPassword = exports.NewPassword = void 0;
/* DO NOT TOUCH THIS AUTO-GENERATED FILE */
/* eslint-disable max-len,@typescript-eslint/no-redeclare */
var connector_1 = require("./connector");
exports.ListAllResources = connector_1.Message.create({
    method: 'GET',
    path: '/',
    status: [200],
});
exports.ApiVersion = connector_1.Message.create({
    method: 'GET',
    path: '/version',
    status: [200],
});
exports.Specification = connector_1.Message.create({
    method: 'GET',
    path: '/spec',
    status: [200],
});
exports.GetBloomFilter = connector_1.Message.create({
    method: 'GET',
    path: '/bloomfilter?rules=false',
    status: [200],
});
exports.GetBloomFilterExpirations = connector_1.Message.create({
    method: 'GET',
    path: '/bloomfilter/expirations',
    status: [200],
});
exports.DeleteBloomFilter = connector_1.Message.create({
    method: 'DELETE',
    path: '/bloomfilter?flush=true',
    status: [204],
});
exports.GetOrestesConfig = connector_1.Message.create({
    method: 'GET',
    path: '/config',
    status: [200],
});
exports.UpdateOrestesConfig = connector_1.Message.create({
    method: 'PUT',
    path: '/config',
    status: [200, 202],
});
exports.Connect = connector_1.Message.create({
    method: 'GET',
    path: '/connect',
    status: [200],
});
exports.Status = connector_1.Message.create({
    method: 'GET',
    path: '/status',
    status: [200],
});
exports.EventsUrl = connector_1.Message.create({
    method: 'GET',
    path: '/events-url',
    status: [200],
});
exports.BannedIp = connector_1.Message.create({
    method: 'GET',
    path: '/banned/:ip',
    status: [204],
});
exports.Banned = connector_1.Message.create({
    method: 'GET',
    path: '/banned',
    status: [],
});
exports.Unban = connector_1.Message.create({
    method: 'DELETE',
    path: '/banned',
    status: [204],
});
exports.UnbanIp = connector_1.Message.create({
    method: 'DELETE',
    path: '/banned/:ip',
    status: [204],
});
exports.GetBucketNames = connector_1.Message.create({
    method: 'GET',
    path: '/db',
    status: [200],
});
exports.GetBucketIds = connector_1.Message.create({
    method: 'GET',
    path: '/db/:bucket/ids?start=0&count=-1',
    status: [200],
});
exports.ExportBucket = connector_1.Message.create({
    method: 'GET',
    path: '/db/:bucket',
    status: [200],
});
exports.ImportBucket = connector_1.Message.create({
    method: 'PUT',
    path: '/db/:bucket',
    status: [200],
});
exports.TruncateBucket = connector_1.Message.create({
    method: 'DELETE',
    path: '/db/:bucket',
    status: [200],
});
exports.CreateObject = connector_1.Message.create({
    method: 'POST',
    path: '/db/:bucket',
    status: [201, 202],
});
exports.GetObject = connector_1.Message.create({
    method: 'GET',
    path: '/db/:bucket/:oid',
    status: [200, 304],
});
exports.ReplaceObject = connector_1.Message.create({
    method: 'PUT',
    path: '/db/:bucket/:oid',
    status: [200, 202],
});
exports.DeleteObject = connector_1.Message.create({
    method: 'DELETE',
    path: '/db/:bucket/:oid',
    status: [202, 204],
});
exports.GetAllSchemas = connector_1.Message.create({
    method: 'GET',
    path: '/schema',
    status: [200],
});
exports.UpdateAllSchemas = connector_1.Message.create({
    method: 'POST',
    path: '/schema',
    status: [200],
});
exports.ReplaceAllSchemas = connector_1.Message.create({
    method: 'PUT',
    path: '/schema',
    status: [200],
});
exports.GetSchema = connector_1.Message.create({
    method: 'GET',
    path: '/schema/:bucket',
    status: [200],
});
exports.UpdateSchema = connector_1.Message.create({
    method: 'POST',
    path: '/schema/:bucket',
    status: [200],
});
exports.ReplaceSchema = connector_1.Message.create({
    method: 'PUT',
    path: '/schema/:bucket',
    status: [200],
});
exports.DeleteSchema = connector_1.Message.create({
    method: 'DELETE',
    path: '/schema/:bucket',
    status: [204],
});
exports.AdhocQuery = connector_1.Message.create({
    method: 'GET',
    path: '/db/:bucket/query?q&start=0&count=-1&sort=&eager=&hinted=',
    status: [200],
});
exports.AdhocQueryPOST = connector_1.Message.create({
    method: 'POST',
    path: '/db/:bucket/query?start=0&count=-1&sort=',
    status: [200],
});
exports.AdhocCountQuery = connector_1.Message.create({
    method: 'GET',
    path: '/db/:bucket/count?q',
    status: [200],
});
exports.AdhocCountQueryPOST = connector_1.Message.create({
    method: 'POST',
    path: '/db/:bucket/count',
    status: [200],
});
exports.ListQueryResources = connector_1.Message.create({
    method: 'GET',
    path: '/query',
    status: [200],
});
exports.CreateQuery = connector_1.Message.create({
    method: 'POST',
    path: '/query',
    status: [201],
});
exports.ListThisQueryResources = connector_1.Message.create({
    method: 'GET',
    path: '/query/:qid',
    status: [200],
});
exports.GetQueryCode = connector_1.Message.create({
    method: 'GET',
    path: '/query/:qid/source',
    status: [200],
});
exports.RunQuery = connector_1.Message.create({
    method: 'GET',
    path: '/query/:qid/result?start=0&count=-1',
    status: [200],
});
exports.GetQueryParameters = connector_1.Message.create({
    method: 'GET',
    path: '/query/:qid/parameters',
    status: [200],
});
exports.NewTransaction = connector_1.Message.create({
    method: 'POST',
    path: '/transaction',
    status: [201],
});
exports.CommitTransaction = connector_1.Message.create({
    method: 'PUT',
    path: '/transaction/:tid/committed',
    status: [200],
});
exports.UpdatePartially = connector_1.Message.create({
    method: 'POST',
    path: '/db/:bucket/:oid',
    status: [200],
});
exports.UpdateField = connector_1.Message.create({
    method: 'POST',
    path: '/db/:bucket/:oid/:field',
    status: [200],
});
exports.Login = connector_1.Message.create({
    method: 'POST',
    path: '/db/User/login',
    status: [200],
});
exports.Register = connector_1.Message.create({
    method: 'POST',
    path: '/db/User/register',
    status: [200, 204],
});
exports.Me = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/me',
    status: [200],
});
exports.ValidateUser = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/validate',
    status: [200],
});
exports.Logout = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/logout',
    status: [204],
});
exports.NewPassword = connector_1.Message.create({
    method: 'POST',
    path: '/db/User/password',
    status: [200],
});
exports.ResetPassword = connector_1.Message.create({
    method: 'POST',
    path: '/db/User/reset',
    status: [200],
});
exports.Verify = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/verify?token=',
    status: [204],
});
exports.ChangeUsername = connector_1.Message.create({
    method: 'POST',
    path: '/db/User/changeUsername',
    status: [204],
});
exports.VerifyUsername = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/verifyUsername?token=',
    status: [204],
});
exports.OAuth2 = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/OAuth/:provider?device_code=&state=&code=&oauth_verifier=&oauth_token=&error_description=',
    status: [200],
});
exports.OAuth1 = connector_1.Message.create({
    method: 'GET',
    path: '/db/User/OAuth1/:provider',
    status: [200],
});
exports.UserToken = connector_1.Message.create({
    method: 'POST',
    path: '/db/User/:oid/token',
    status: [200],
});
exports.RevokeUserToken = connector_1.Message.create({
    method: 'DELETE',
    path: '/db/User/:oid/token',
    status: [204],
});
exports.GetBaqendCode = connector_1.Message.create({
    method: 'GET',
    path: '/code/:bucket/:type',
    status: [200],
});
exports.SetBaqendCode = connector_1.Message.create({
    method: 'PUT',
    path: '/code/:bucket/:type',
    status: [200, 202],
});
exports.DeleteBaqendCode = connector_1.Message.create({
    method: 'DELETE',
    path: '/code/:bucket/:type',
    status: [202, 204],
});
exports.PostBaqendModule = connector_1.Message.create({
    method: 'POST',
    path: '/code/:bucket',
    status: [200, 204],
});
exports.GetBaqendModule = connector_1.Message.create({
    method: 'GET',
    path: '/code/:bucket',
    status: [200, 204],
});
exports.GetAllModules = connector_1.Message.create({
    method: 'GET',
    path: '/code',
    status: [200],
});
exports.ListFiles = connector_1.Message.create({
    method: 'GET',
    path: '/file/:bucket/ids?path=/&start=&count=-1&deep=false',
    status: [200],
});
exports.ListBuckets = connector_1.Message.create({
    method: 'GET',
    path: '/file/buckets',
    status: [200],
});
exports.DownloadArchive = connector_1.Message.create({
    method: 'GET',
    path: '/file',
    status: [200],
});
exports.UploadPatchArchive = connector_1.Message.create({
    method: 'POST',
    path: '/file',
    status: [200],
});
exports.GetFileBucketMetadata = connector_1.Message.create({
    method: 'GET',
    path: '/file/:bucket',
    status: [200],
});
exports.SetFileBucketMetadata = connector_1.Message.create({
    method: 'PUT',
    path: '/file/:bucket',
    status: [204],
});
exports.DeleteFileBucket = connector_1.Message.create({
    method: 'DELETE',
    path: '/file/:bucket',
    status: [204],
});
exports.CreateFile = connector_1.Message.create({
    method: 'POST',
    path: '/file/:bucket',
    status: [200],
});
exports.DownloadFile = connector_1.Message.create({
    method: 'GET',
    path: '/file/:bucket/*oid',
    status: [200, 304],
});
exports.UploadFile = connector_1.Message.create({
    method: 'PUT',
    path: '/file/:bucket/*oid',
    status: [200],
});
exports.GetFileMetadata = connector_1.Message.create({
    method: 'HEAD',
    path: '/file/:bucket/*oid',
    status: [200],
});
exports.UpdateFileMetadata = connector_1.Message.create({
    method: 'POST',
    path: '/file/:bucket/*oid',
    status: [200],
});
exports.DeleteFile = connector_1.Message.create({
    method: 'DELETE',
    path: '/file/:bucket/*oid',
    status: [200, 204],
});
exports.CreateManifest = connector_1.Message.create({
    method: 'POST',
    path: '/pwa/manifest',
    status: [202],
});
exports.DownloadAsset = connector_1.Message.create({
    method: 'GET',
    path: '/asset/*url',
    status: [200, 304],
});
exports.RevalidateAssets = connector_1.Message.create({
    method: 'POST',
    path: '/asset/revalidate',
    status: [202],
});
exports.GetRevalidationStatus = connector_1.Message.create({
    method: 'GET',
    path: '/asset/revalidate/:id',
    status: [200, 202],
});
exports.CancelRevalidation = connector_1.Message.create({
    method: 'DELETE',
    path: '/asset/revalidate/:id',
    status: [202],
});
exports.GetAllRevalidationStatus = connector_1.Message.create({
    method: 'GET',
    path: '/asset/revalidate?state=',
    status: [200],
});
exports.CleanUpAssets = connector_1.Message.create({
    method: 'POST',
    path: '/asset/cleanup',
    status: [202],
});
exports.CleanUpStorage = connector_1.Message.create({
    method: 'POST',
    path: '/asset/cleanup/storage',
    status: [202],
});
exports.ListIndexes = connector_1.Message.create({
    method: 'GET',
    path: '/index/:bucket',
    status: [200],
});
exports.CreateDropIndex = connector_1.Message.create({
    method: 'POST',
    path: '/index/:bucket',
    status: [202],
});
exports.DropAllIndexes = connector_1.Message.create({
    method: 'DELETE',
    path: '/index/:bucket',
    status: [202],
});
exports.DeviceRegister = connector_1.Message.create({
    method: 'POST',
    path: '/db/Device/register',
    status: [200],
});
exports.DevicePush = connector_1.Message.create({
    method: 'POST',
    path: '/db/Device/push',
    status: [204],
});
exports.DeviceRegistered = connector_1.Message.create({
    method: 'GET',
    path: '/db/Device/registered',
    status: [200],
});
exports.VAPIDKeys = connector_1.Message.create({
    method: 'POST',
    path: '/config/VAPIDKeys',
    status: [200],
});
exports.VAPIDPublicKey = connector_1.Message.create({
    method: 'GET',
    path: '/config/VAPIDPublicKey',
    status: [200],
});
exports.GCMAKey = connector_1.Message.create({
    method: 'POST',
    path: '/config/GCMKey',
    status: [204],
});
exports.UploadAPNSCertificate = connector_1.Message.create({
    method: 'POST',
    path: '/config/APNSCert',
    status: [204],
});
exports.Install = connector_1.Message.create({
    method: 'GET',
    path: '/speedkit?d=',
    status: [200],
});
exports.SW = connector_1.Message.create({
    method: 'GET',
    path: '/speedkit?r=&v=',
    status: [200],
});
exports.Beacon = connector_1.Message.create({
    method: 'POST',
    path: '/rum/pi',
    status: [200],
});
