// baqend/commonjs/lib/connector/FetchConnector.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchConnector = void 0;
var Connector_1 = require("./Connector");
var FetchConnector = /** @class */ (function (_super) {
    __extends(FetchConnector, _super);
    function FetchConnector() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Indicates if this connector implementation is usable for the given host and port
     * @return
     */
    FetchConnector.isUsable = function () {
        return typeof fetch !== 'undefined';
    };
    /**
     * @inheritDoc
     */
    FetchConnector.prototype.doSend = function (message, request, receive) {
        var url = this.origin + this.basePath + request.path;
        var method = request.method;
        var headers = request.headers;
        var entity = request.entity;
        var credentials = message.withCredentials ? 'include' : 'same-origin';
        return fetch(url, {
            method: method,
            headers: headers,
            body: entity,
            credentials: credentials,
        }).then(function (res) {
            var responseHeaders = {};
            Connector_1.Connector.RESPONSE_HEADERS.forEach(function (name) {
                responseHeaders[name] = res.headers.get ? res.headers.get(name) : res.headers[name];
            });
            var response = {
                headers: responseHeaders,
                status: res.status,
                entity: res,
            };
            receive(response);
        });
    };
    /**
     * @inheritDoc
     */
    FetchConnector.prototype.fromFormat = function (response, rawEntity, type) {
        if (type === 'json') {
            return rawEntity.json();
        }
        if (type === 'blob') {
            return rawEntity.blob();
        }
        if (type === 'data-url' || type === 'base64') {
            return rawEntity.blob().then(function (entity) {
                var reader = new FileReader();
                reader.readAsDataURL(entity);
                return new Promise(function (resolve, reject) {
                    reader.onload = resolve;
                    reader.onerror = reject;
                }).then(function () {
                    var result = reader.result;
                    if (type === 'base64') {
                        result = result.substring(result.indexOf(',') + 1);
                    }
                    return result;
                });
            });
        }
        return rawEntity;
    };
    /**
     * @inheritDoc
     */
    FetchConnector.prototype.toFormat = function (message) {
        var type = message.request.type;
        if (type) {
            var entity = message.request.entity;
            var mimeType = message.mimeType();
            switch (type) {
                case 'blob':
                    mimeType = mimeType || entity.type;
                    break;
                case 'arraybuffer':
                case 'form':
                    break;
                case 'data-url': {
                    var match = entity.match(/^data:(.+?)(;base64)?,(.*)$/);
                    var isBase64 = match[2];
                    // eslint-disable-next-line prefer-destructuring
                    entity = match[3];
                    type = 'blob';
                    mimeType = mimeType || match[1];
                    if (!isBase64) {
                        entity = decodeURIComponent(entity);
                        break;
                    }
                } // fallthrough
                case 'base64': {
                    var binaryStr = atob(entity);
                    var len = binaryStr.length;
                    var array = new Uint8Array(len);
                    for (var i = 0; i < len; i += 1) {
                        array[i] = binaryStr.charCodeAt(i);
                    }
                    type = 'blob';
                    entity = new Blob([array], { type: mimeType });
                    break;
                }
                case 'json': {
                    if (typeof entity !== 'string') {
                        entity = JSON.stringify(entity);
                    }
                    break;
                }
                case 'text':
                    break;
                default:
                    throw new Error("Supported request format:" + type);
            }
            message.entity(entity, type).mimeType(mimeType);
        }
    };
    return FetchConnector;
}(Connector_1.Connector));
exports.FetchConnector = FetchConnector;
Connector_1.Connector.connectors.push(FetchConnector);
