// baqend/commonjs/lib/connector/IFrameConnector.js
"use strict";
/* this connector will only be choose in browser compatible environments */
/* eslint no-restricted-globals: ["off", "location", "addEventListener"] */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.IFrameConnector = void 0;
var Connector_1 = require("./Connector");
var XMLHttpConnector_1 = require("./XMLHttpConnector");
var IFrameConnector = /** @class */ (function (_super) {
    __extends(IFrameConnector, _super);
    function IFrameConnector(host, port, secure, basePath) {
        var _this = _super.call(this, host, port, secure, basePath) || this;
        _this.queue = null;
        _this.connected = false;
        _this.mid = 0;
        _this.messages = {};
        _this.doReceive = _this.doReceive.bind(_this);
        addEventListener('message', _this.doReceive, false);
        return _this;
    }
    /**
     * Indicates if this connector implementation is usable for the given host and port
     * @param host
     * @param port
     * @param secure
     * @return
     */
    IFrameConnector.isUsable = function (host, port, secure) {
        // we use location directly here, since there exists environments, which provide a location and a document but
        // no window object
        if (typeof location === 'undefined' || typeof document === 'undefined') {
            return false;
        }
        var locationSecure = location.protocol === 'https:';
        var locationPort = location.port || (locationSecure ? 443 : 80);
        return location.hostname !== host || locationPort !== port || locationSecure !== secure;
    };
    IFrameConnector.prototype.load = function (path) {
        this.iframe = document.createElement('iframe');
        this.iframe.src = this.origin + this.basePath + path;
        this.iframe.setAttribute('style', IFrameConnector.style);
        document.body.appendChild(this.iframe);
        this.queue = [];
        this.iframe.addEventListener('load', this.onLoad.bind(this), false);
    };
    IFrameConnector.prototype.onLoad = function () {
        if (!this.queue) {
            return;
        }
        var queue = this.queue;
        for (var i = 0; i < queue.length; i += 1) {
            this.postMessage(queue[i]);
        }
        this.queue = null;
    };
    /**
     * @inheritDoc
     */
    IFrameConnector.prototype.doSend = function (message, request, receive) {
        var _this = this;
        // binary data will be send and received directly
        if (message.isBinary) {
            _super.prototype.doSend.call(this, message, request, receive);
            return;
        }
        if (!this.iframe) {
            this.load(message.request.path);
            // ensure that we get a local resource cache hit
            // eslint-disable-next-line no-param-reassign
            message.request.path = '/connect';
        }
        var msg = {
            mid: this.mid += 1,
            method: request.method,
            path: request.path,
            headers: request.headers,
            entity: request.entity,
            responseHeaders: Connector_1.Connector.RESPONSE_HEADERS,
        };
        this.messages[msg.mid] = receive;
        var strMsg = JSON.stringify(msg);
        if (this.queue) {
            this.queue.push(strMsg);
        }
        else {
            this.postMessage(strMsg);
        }
        if (!this.connected) {
            setTimeout(function () {
                if (_this.messages[msg.mid]) {
                    delete _this.messages[msg.mid];
                    receive({
                        status: 0,
                        error: new Error('Connection refused.'),
                        headers: {},
                    });
                }
            }, 10000);
        }
    };
    IFrameConnector.prototype.postMessage = function (msg) {
        this.iframe.contentWindow.postMessage(msg, this.origin);
    };
    IFrameConnector.prototype.doReceive = function (event) {
        if (event.origin !== this.origin || event.data[0] !== '{') {
            return;
        }
        var msg = JSON.parse(event.data);
        var receive = this.messages[msg.mid];
        if (receive) {
            delete this.messages[msg.mid];
            this.connected = true;
            receive({
                status: msg.status,
                headers: msg.headers,
                entity: msg.entity,
            });
        }
    };
    IFrameConnector.style = 'width:1px;height:1px;position:absolute;top:-10px;left:-10px;';
    return IFrameConnector;
}(XMLHttpConnector_1.XMLHttpConnector));
exports.IFrameConnector = IFrameConnector;
Connector_1.Connector.connectors.push(IFrameConnector);
