// baqend/commonjs/lib/connector/NodeConnector.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeConnector = void 0;
var https_1 = __importDefault(require("https"));
var http_1 = __importDefault(require("http"));
var Connector_1 = require("./Connector");
var error_1 = require("../error");
var NodeConnector = /** @class */ (function (_super) {
    __extends(NodeConnector, _super);
    function NodeConnector(host, port, secure, basePath) {
        var _this = _super.call(this, host, port, secure, basePath) || this;
        _this.cookie = null;
        _this.http = secure ? https_1.default : http_1.default;
        return _this;
    }
    NodeConnector.isUsable = function () {
        // prevent using when it is shimmed
        return !!(http_1.default && http_1.default.Server);
    };
    /**
     * @inheritDoc
     */
    NodeConnector.prototype.doSend = function (message, request, receive) {
        var _this = this;
        var entity = request.entity;
        var type = request.type;
        var responseType = message.responseType();
        if (this.cookie && message.withCredentials) {
            request.headers.cookie = this.cookie;
        }
        var nodeRequest = __assign(__assign({}, request), { host: this.host, port: this.port, path: this.basePath + request.path });
        var req = this.http.request(nodeRequest, function (res) {
            var cookie = res.headers['set-cookie'];
            if (cookie) {
                // cookie may be an array, convert it to a string
                _this.cookie = _this.parseCookie("" + cookie);
            }
            var status = res.statusCode || 0;
            if (status >= 400) {
                responseType = 'json';
            }
            if (responseType === 'stream') {
                receive({
                    status: status,
                    headers: res.headers,
                    entity: res,
                });
                return;
            }
            var binary = responseType && responseType !== 'text' && responseType !== 'json';
            var chunks = [];
            if (!binary) {
                res.setEncoding('utf-8');
            }
            res.on('data', function (chunk) {
                chunks.push(chunk);
            });
            res.on('end', function () {
                receive({
                    status: status,
                    headers: res.headers,
                    entity: binary ? Buffer.concat(chunks) : chunks.join(''),
                });
            });
        });
        req.on('error', function (e) {
            receive({
                status: 0,
                headers: {},
                error: e,
            });
        });
        if (type === 'stream') {
            entity.pipe(req);
        }
        else if (type === 'buffer') {
            req.end(entity);
        }
        else if (type) {
            req.end(entity, 'utf8');
        }
        else {
            req.end();
        }
    };
    /**
     * Parse the cookie header
     * @param header
     * @return
     */
    NodeConnector.prototype.parseCookie = function (header) {
        var parts = header.split(';');
        for (var i = 0, len = parts.length; i < len; i += 1) {
            var part = parts[i];
            if (part.trim().indexOf('Expires=') === 0) {
                var date = Date.parse(part.substring(8));
                if (date < Date.now()) {
                    return null;
                }
            }
        }
        return parts[0];
    };
    /**
     * @inheritDoc
     */
    NodeConnector.prototype.toFormat = function (message) {
        var type = message.request.type;
        if (type) {
            var entity = message.request.entity;
            var mimeType = message.mimeType();
            switch (type) {
                case 'stream':
                    if (!message.contentLength()) {
                        throw new error_1.PersistentError('You must specify a content length while making a stream based upload.');
                    }
                    break;
                case 'buffer':
                    break;
                case 'arraybuffer':
                    type = 'buffer';
                    entity = Buffer.from(entity);
                    break;
                case 'data-url': {
                    var match = entity.match(/^data:(.+?)(;base64)?,(.*)$/);
                    var isBase64 = match[2];
                    // eslint-disable-next-line prefer-destructuring
                    entity = match[3];
                    type = 'buffer';
                    mimeType = mimeType || match[1];
                    if (isBase64) {
                        entity = Buffer.from(entity, 'base64');
                    }
                    else {
                        entity = Buffer.from(decodeURIComponent(entity), 'utf8');
                    }
                    break;
                }
                case 'base64':
                    type = 'buffer';
                    entity = Buffer.from(entity, 'base64');
                    break;
                case 'json':
                    if (typeof entity !== 'string') {
                        entity = JSON.stringify(entity);
                    }
                    break;
                case 'text':
                    break;
                default:
                    throw new Error("The request type " + type + " is not supported");
            }
            message.entity(entity, type).mimeType(mimeType);
        }
    };
    /**
     * @inheritDoc
     */
    NodeConnector.prototype.fromFormat = function (response, entity, type) {
        switch (type) {
            case 'json':
                return JSON.parse(entity);
            case 'data-url':
            case 'base64': {
                var base64 = entity.toString('base64');
                if (type === 'base64') {
                    return base64;
                }
                return "data:" + response.headers['content-type'] + ";base64," + base64;
            }
            case 'arraybuffer':
                return entity.buffer.slice(entity.byteOffset, entity.byteOffset + entity.byteLength);
            default:
                return entity;
        }
    };
    return NodeConnector;
}(Connector_1.Connector));
exports.NodeConnector = NodeConnector;
Connector_1.Connector.connectors.push(NodeConnector);
