// baqend/commonjs/lib/connector/WebSocketConnector.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketConnector = void 0;
var error_1 = require("../error");
var websocket_1 = require("../util/websocket");
var observable_1 = require("../util/observable");
var WebSocketConnector = /** @class */ (function () {
    /**
     *url
     */
    function WebSocketConnector(url) {
        this.observers = {};
        this.socket = null;
        this.url = url;
    }
    /**
     *url The websocket connect script url
     *a websocket connection
     */
    WebSocketConnector.create = function (url) {
        var websocket = this.websockets[url];
        if (!websocket) {
            websocket = new WebSocketConnector(url);
            this.websockets[url] = websocket;
        }
        return websocket;
    };
    WebSocketConnector.prototype.open = function () {
        var _this = this;
        if (!this.socket) {
            var socket_1 = new websocket_1.WebSocket(this.url);
            var socketPromise_1;
            var handleSocketCompletion_1 = function (error) {
                // observable error calls can throw an exception therefore cleanup beforehand
                var isError = false;
                if (_this.socket === socketPromise_1) {
                    isError = socket_1.readyState !== 3;
                    _this.socket = null;
                }
                var firstErr = null;
                Object.keys(_this.observers).forEach(function (id) {
                    var observer = _this.observers[id];
                    delete _this.observers[id]; // unsubscribe to allow re subscriptions
                    if (!observer) {
                        return;
                    }
                    try {
                        if (isError) {
                            observer.error(new error_1.CommunicationError(null, __assign({ status: 0, headers: {} }, (error instanceof Error && { error: error }))));
                        }
                        else {
                            observer.complete();
                        }
                    }
                    catch (e) {
                        if (!firstErr) {
                            firstErr = e;
                        }
                    }
                });
                if (firstErr) {
                    throw firstErr;
                }
            };
            socket_1.onerror = handleSocketCompletion_1;
            socket_1.onclose = handleSocketCompletion_1;
            socket_1.onmessage = function (event) {
                var message = JSON.parse(event.data);
                message.date = new Date(message.date);
                var id = message.id;
                if (!id) {
                    if (message.type === 'error') {
                        handleSocketCompletion_1(message);
                    }
                    return;
                }
                var observer = _this.observers[id];
                if (observer) {
                    if (message.type === 'error') {
                        observer.error(new error_1.CommunicationError(null, message));
                    }
                    else {
                        observer.next(message);
                    }
                }
            };
            socketPromise_1 = new Promise(function (resolve) {
                socket_1.onopen = function () { return resolve(socket_1); };
            });
            this.socket = socketPromise_1;
        }
        return this.socket;
    };
    WebSocketConnector.prototype.close = function () {
        if (this.socket) {
            this.socket.then(function (socket) {
                socket.close();
            });
            this.socket = null;
        }
    };
    /**
     *tokenStorage
     *id subscription ID
     *The channel for sending and receiving messages
     */
    WebSocketConnector.prototype.openStream = function (tokenStorage, id) {
        var _this = this;
        var stream = new observable_1.Observable(function (observer) {
            if (_this.observers[id]) {
                throw new Error('Only one subscription per stream is allowed.');
            }
            _this.observers[id] = observer;
            return function () {
                // cleanup only our subscription and handle re subscription on the same stream id correctly
                if (_this.observers[id] === observer) {
                    delete _this.observers[id];
                }
            };
        });
        Object.assign(stream, {
            send: function (message) {
                _this.open().then(function (socket) {
                    var jsonMessage = JSON.stringify(__assign(__assign({ id: id }, message), (tokenStorage.token && { token: tokenStorage.token })));
                    socket.send(jsonMessage);
                });
            },
        });
        return stream;
    };
    /**
     * Map of all available connectors to their respective websocket connections
     */
    WebSocketConnector.websockets = {};
    return WebSocketConnector;
}());
exports.WebSocketConnector = WebSocketConnector;
