// baqend/commonjs/lib/intersection/Validator.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validator = void 0;
var FallBachValLib = {};
var valLib = FallBachValLib;
try {
    // we load this module as an optional external dependency
    // eslint-disable-next-line global-require
    valLib = require('validator');
}
catch (e) {
    // ignore loading optional module error
}
var Validator = /** @class */ (function () {
    function Validator(key, entity) {
        /**
         * The cached errors of the validation
         */
        this.errors = [];
        this.key = key;
        this.entity = entity;
    }
    /**
     * Compiles the given validation code for the managedType
     * @param managedType The managedType of the code
     * @param validationCode The validation code
     * @return the parsed validation function
     */
    Validator.compile = function (managedType, validationCode) {
        var keys = [];
        var iter = managedType.attributes();
        for (var el = iter.next(); !el.done; el = iter.next()) {
            var attr = el.value;
            keys.push(attr.name);
        }
        // eslint-disable-next-line @typescript-eslint/no-implied-eval,no-new-func
        var fn = new (Function.bind.apply(Function, __spreadArray(__spreadArray([void 0], keys), [validationCode])))();
        return function onValidate(argObj) {
            if (valLib === FallBachValLib) {
                throw new Error('Validation code will not be executed. Make sure that the validator package is correctly provided as an external dependency.');
            }
            var args = keys.map(function (name) { return argObj[name]; });
            return fn.apply({}, args);
        };
    };
    Object.defineProperty(Validator.prototype, "value", {
        /**
         * Gets the value of the attribute
         * @return Value
         */
        get: function () {
            return this.entity[this.key];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Validator.prototype, "isValid", {
        /**
         * Checks if the attribute is valid
         * @return
         */
        get: function () {
            return this.errors.length === 0;
        },
        enumerable: false,
        configurable: true
    });
    Validator.prototype.is = function (error, fn) {
        if (error instanceof Function) {
            return this.is('is', error);
        }
        if (fn(this.value, valLib) === false) {
            this.errors.push(error);
        }
        return this;
    };
    Validator.prototype.callMethod = function (method, errorMessage, argumentList) {
        var args = argumentList || [];
        try {
            args.unshift(this.toStringValue());
            if (valLib[method].apply(this, args) === false) {
                this.errors.push(errorMessage || method);
            }
        }
        catch (e) {
            this.errors.push(errorMessage || e.message);
        }
        return this;
    };
    Validator.prototype.toStringValue = function () {
        var value = this.value;
        if (typeof value === 'string' || value instanceof Date) {
            return value;
        }
        return JSON.stringify(value);
    };
    Validator.prototype.toJSON = function () {
        return {
            isValid: this.isValid,
            errors: this.errors,
        };
    };
    return Validator;
}());
exports.Validator = Validator;
var OTHER_VALIDATORS = ['contains', 'equals', 'matches'];
Object.keys(valLib).forEach(function (name) {
    if (name.startsWith('is') || OTHER_VALIDATORS.includes(name)) {
        // use function here to keep the correct this context
        Validator.prototype[name] = function validate() {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var error = typeof args[0] === 'string' ? args.shift() : null;
            return this.callMethod(name, error, args);
        };
    }
});
