// baqend/commonjs/lib/EntityManagerFactory.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityManagerFactory = void 0;
var EntityManager_1 = require("./EntityManager");
var message = __importStar(require("./message"));
var connector_1 = require("./connector");
var util_1 = require("./util");
var intersection_1 = require("./intersection");
var metamodel_1 = require("./metamodel");
var CONNECTED = Symbol('Connected');
var WS = Symbol('WebSocket');
var EntityManagerFactory = /** @class */ (function (_super) {
    __extends(EntityManagerFactory, _super);
    /**
     * Creates a new EntityManagerFactory connected to the given destination
     * @param [options] The destination to connect with, or an options object
     * @param [options.host] The destination to connect with
     * @param [options.port=80|443] The optional destination port to connect with
     * @param [options.secure=false] <code>true</code> To use a secure ssl encrypted connection
     * @param [options.basePath="/v1"] The base path of the api
     * @param [options.schema=null] The serialized schema as json used to initialize the metamodel
     * @param [options.tokenStorage] The tokenStorage which should be used by this emf
     * @param [options.tokenStorageFactory] The tokenStorage factory implementation which
     * should be used for token storage
     * @param [options.staleness=60] The maximum staleness of objects that are acceptable while reading cached
     * data
     */
    function EntityManagerFactory(options) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this) || this;
        _this.connection = null;
        _this.metamodel = _this.createMetamodel();
        _this.code = new intersection_1.Code(_this.metamodel, _this);
        _this.tokenStorageFactory = intersection_1.TokenStorage.WEB_STORAGE || intersection_1.TokenStorage.GLOBAL;
        _this.tokenStorage = null;
        _this.staleness = null;
        var opt = typeof options === 'string' ? { host: options } : options || {};
        _this.configure(opt);
        var isReady = true;
        var ready = new Promise(function (success) {
            _this[CONNECTED] = success;
        });
        if (opt.host) {
            _this.connect(opt.host, opt.port, opt.secure, opt.basePath);
        }
        else {
            isReady = false;
        }
        if (!_this.tokenStorage) {
            isReady = false;
            ready = ready
                .then(function () { return _this.tokenStorageFactory.create(_this.connection.origin); })
                .then(function (tokenStorage) {
                _this.tokenStorage = tokenStorage;
            });
        }
        if (opt.schema) {
            _this.connectData = opt;
            _this.metamodel.init(opt.schema);
        }
        else {
            isReady = false;
            ready = ready.then(function () {
                var msg = new message.Connect();
                msg.withCredentials = true; // used for registered devices
                if (_this.staleness === 0) {
                    msg.noCache();
                }
                return _this.send(msg);
            }).then(function (response) {
                _this.connectData = response.entity;
                if (_this.staleness === null) {
                    _this.staleness = _this.connectData.bloomFilterRefresh || 60;
                }
                if (!_this.metamodel.isInitialized) {
                    _this.metamodel.init(_this.connectData.schema);
                }
                _this.tokenStorage.update(_this.connectData.token);
            });
        }
        if (!isReady) {
            _this.withLock(function () { return ready; }, true);
        }
        return _this;
    }
    Object.defineProperty(EntityManagerFactory.prototype, "websocket", {
        /**
         * Retrieves the websocket connection if the websocket SDK is available
         */
        get: function () {
            if (!this[WS]) {
                var secure = this.connection.secure;
                var url = void 0;
                if (this.connectData.websocket) {
                    url = (secure ? 'wss:' : 'ws:') + this.connectData.websocket;
                }
                else {
                    url = this.connection.origin.replace(/^http/, 'ws') + this.connection.basePath + "/events";
                }
                this[WS] = connector_1.WebSocketConnector.create(url);
            }
            return this[WS];
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Apply additional configurations to this EntityManagerFactory
     * @param options The additional configuration options
     * @param [options.tokenStorage] The tokenStorage which should be used by this emf
     * @param [options.tokenStorageFactory] The tokenStorage factory implementation which
     * should be used for token storage
     * @param [options.staleness=60] The maximum staleness of objects that are acceptable while reading cached
     * data, <code>0</code> to always bypass the browser cache
     * @return
     */
    EntityManagerFactory.prototype.configure = function (options) {
        if (this.connection) {
            throw new Error('The EntityManagerFactory can only be configured before is is connected.');
        }
        if (options.tokenStorage) {
            this.tokenStorage = options.tokenStorage;
        }
        if (options.tokenStorageFactory) {
            this.tokenStorageFactory = options.tokenStorageFactory;
        }
        if (options.staleness !== undefined) {
            this.staleness = options.staleness;
        }
    };
    EntityManagerFactory.prototype.connect = function (hostOrApp, port, secure, basePath) {
        if (this.connection) {
            throw new Error('The EntityManagerFactory is already connected.');
        }
        if (typeof port === 'boolean') {
            return this.connect(hostOrApp, 0, port, secure);
        }
        this.connection = connector_1.Connector.create(hostOrApp, port, secure, basePath);
        this[CONNECTED]();
        return this.ready();
    };
    /**
     * Creates a new Metamodel instance, which is not connected
     * @return A new Metamodel instance
     */
    EntityManagerFactory.prototype.createMetamodel = function () {
        return new metamodel_1.Metamodel(this);
    };
    /**
     * Create a new application-managed EntityManager.
     *
     * @param useSharedTokenStorage The token storage to persist the authorization token, or
     * <code>true</code> To use the shared token storage of the emf.
     * <code>false</code> To use a instance based storage.
     *
     * @return a new entityManager
     */
    EntityManagerFactory.prototype.createEntityManager = function (useSharedTokenStorage) {
        var _this = this;
        var em = new EntityManager_1.EntityManager(this);
        if (this.isReady) {
            em.connected(this.connection, this.connectData, useSharedTokenStorage ? this.tokenStorage : new intersection_1.TokenStorage(this.connection.origin));
        }
        else {
            em.withLock(function () { return _this.ready().then(function () {
                em.connected(_this.connection, _this.connectData, useSharedTokenStorage ? _this.tokenStorage : new intersection_1.TokenStorage(_this.connection.origin));
            }); }, true);
        }
        return em;
    };
    EntityManagerFactory.prototype.send = function (msg) {
        if (!msg.tokenStorage()) {
            msg.tokenStorage(this.tokenStorage);
        }
        return this.connection.send(msg);
    };
    return EntityManagerFactory;
}(util_1.Lockable));
exports.EntityManagerFactory = EntityManagerFactory;
