// baqend/commonjs/lib/error/PersistentError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PersistentError = void 0;
// eslint-disable-next-line @typescript-eslint/no-redeclare
exports.PersistentError = (function () {
    function PersistentErrorConstructor(message, cause) {
        if (Object.prototype.hasOwnProperty.call(Error, 'captureStackTrace')) {
            Error.captureStackTrace(this, this.constructor);
        }
        else {
            this.stack = (new Error()).stack;
        }
        this.message = (message || 'An unexpected persistent error occurred.');
        this.name = this.constructor.name;
        if (cause) {
            this.cause = cause;
            if (cause.stack) {
                this.stack += "\nCaused By: " + cause.stack;
            }
        }
    }
    // custom errors must be manually extended, since JS Errors can't be super called in a class hierarchy,
    // otherwise the super call destroys the origin 'this' reference
    PersistentErrorConstructor.prototype = Object.create(Error.prototype, {
        constructor: {
            value: PersistentErrorConstructor,
            writable: true,
            enumerable: false,
            configurable: true,
        },
    });
    return PersistentErrorConstructor;
})();
exports.PersistentError.of = function of(error) {
    if (error instanceof exports.PersistentError) {
        return error;
    }
    return new exports.PersistentError(null, error);
};
