// baqend/commonjs/lib/query/Condition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Condition = void 0;
var Query_1 = require("./Query");
// eslint-disable-next-line @typescript-eslint/no-redeclare
exports.Condition = {
    where: function (conditions) {
        return this.addFilter(null, null, conditions);
    },
    equal: function (field, value) {
        return this.addFilter(field, null, value);
    },
    notEqual: function (field, value) {
        return this.addFilter(field, '$ne', value);
    },
    greaterThan: function (field, value) {
        return this.addFilter(field, '$gt', value);
    },
    greaterThanOrEqualTo: function (field, value) {
        return this.addFilter(field, '$gte', value);
    },
    lessThan: function (field, value) {
        return this.addFilter(field, '$lt', value);
    },
    lessThanOrEqualTo: function (field, value) {
        return this.addFilter(field, '$lte', value);
    },
    between: function (field, greaterValue, lessValue) {
        return this
            .addFilter(field, '$gt', greaterValue)
            .addFilter(field, '$lt', lessValue);
    },
    in: function (field) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.addFilter(field, '$in', Query_1.flatArgs(args));
    },
    notIn: function (field) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.addFilter(field, '$nin', Query_1.flatArgs(args));
    },
    isNull: function (field) {
        return this.equal(field, null);
    },
    isNotNull: function (field) {
        return this.addFilter(field, '$exists', true)
            .addFilter(field, '$ne', null);
    },
    containsAll: function (field) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return this.addFilter(field, '$all', Query_1.flatArgs(args));
    },
    mod: function (field, divisor, remainder) {
        return this.addFilter(field, '$mod', [divisor, remainder]);
    },
    matches: function (field, regExp) {
        var reg = regExp instanceof RegExp ? regExp : new RegExp(regExp);
        if (reg.ignoreCase) {
            throw new Error('RegExp.ignoreCase flag is not supported.');
        }
        if (reg.global) {
            throw new Error('RegExp.global flag is not supported.');
        }
        if (reg.source.indexOf('^') !== 0) {
            throw new Error('regExp must be an anchored expression, i.e. it must be started with a ^.');
        }
        var result = this.addFilter(field, '$regex', reg.source);
        if (reg.multiline) {
            result.addFilter(field, '$options', 'm');
        }
        return result;
    },
    size: function (field, size) {
        return this.addFilter(field, '$size', size);
    },
    near: function (field, geoPoint, maxDistance) {
        return this.addFilter(field, '$nearSphere', {
            $geometry: {
                type: 'Point',
                coordinates: [geoPoint.longitude, geoPoint.latitude],
            },
            $maxDistance: maxDistance,
        });
    },
    withinPolygon: function (field) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var geoPoints = Query_1.flatArgs(args);
        return this.addFilter(field, '$geoWithin', {
            $geometry: {
                type: 'Polygon',
                coordinates: [geoPoints.map(function (geoPoint) { return [geoPoint.longitude, geoPoint.latitude]; })],
            },
        });
    },
};
// aliases
Object.assign(exports.Condition, {
    eq: exports.Condition.equal,
    ne: exports.Condition.notEqual,
    lt: exports.Condition.lessThan,
    le: exports.Condition.lessThanOrEqualTo,
    gt: exports.Condition.greaterThan,
    ge: exports.Condition.greaterThanOrEqualTo,
    containsAny: exports.Condition.in,
});
