// baqend/commonjs/lib/metamodel/BasicType.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicType = void 0;
var binding_1 = require("../binding");
var GeoPoint_1 = require("../GeoPoint");
var Type_1 = require("./Type");
function dateToJson(value) {
    // remove trailing zeros
    return value instanceof Date ? value.toISOString().replace(/\.?0*Z/, 'Z') : null;
}
function jsonToDate(json, currentValue) {
    var date = typeof json === 'string' ? new Date(json) : null;
    if (currentValue && date) {
        // compare normalized date strings instead of plain strings
        return currentValue.toISOString() === date.toISOString() ? currentValue : date;
    }
    return date;
}
var BasicType = /** @class */ (function (_super) {
    __extends(BasicType, _super);
    /**
     * Creates a new instance of a native db type
     * @param ref The db ref of this type
     * @param typeConstructor The javascript class of this type
     * @param noResolving Indicates if this type is not the main type of the constructor
     */
    function BasicType(ref, typeConstructor, noResolving) {
        var _this = this;
        var id = ref.indexOf('/db/') === 0 ? ref : "/db/" + ref;
        _this = _super.call(this, id, typeConstructor) || this;
        _this.noResolving = !!noResolving;
        return _this;
    }
    Object.defineProperty(BasicType.prototype, "persistenceType", {
        /**
         * @inheritDoc
         */
        get: function () {
            return Type_1.PersistenceType.BASIC;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @inheritDoc
     */
    BasicType.prototype.toJsonValue = function (state, currentValue) {
        return currentValue === null || currentValue === undefined ? null : this.typeConstructor(currentValue);
    };
    /**
     * @inheritDoc
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    BasicType.prototype.fromJsonValue = function (state, json, currentValue) {
        return json === null || json === undefined ? null : json;
    };
    BasicType.prototype.toString = function () {
        return "BasicType(" + this.ref + ")";
    };
    BasicType.Boolean = new /** @class */ (function (_super) {
        __extends(BooleanType, _super);
        function BooleanType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        BooleanType.prototype.fromJsonValue = function (state, json, currentValue) {
            return typeof json === 'string' ? json !== 'false' : _super.prototype.fromJsonValue.call(this, state, json, currentValue);
        };
        return BooleanType;
    }(BasicType))('Boolean', Boolean);
    BasicType.Double = new /** @class */ (function (_super) {
        __extends(DoubleType, _super);
        function DoubleType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        DoubleType.prototype.fromJsonValue = function (state, json, currentValue) {
            return typeof json === 'string' ? parseFloat(json) : _super.prototype.fromJsonValue.call(this, state, json, currentValue);
        };
        return DoubleType;
    }(BasicType))('Double', Number);
    BasicType.Integer = new /** @class */ (function (_super) {
        __extends(IntegerType, _super);
        function IntegerType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        IntegerType.prototype.fromJsonValue = function (state, json, currentValue) {
            return typeof json === 'string' ? parseInt(json, 10) : _super.prototype.fromJsonValue.call(this, state, json, currentValue);
        };
        return IntegerType;
    }(BasicType))('Integer', Number);
    BasicType.String = new /** @class */ (function (_super) {
        __extends(StringType, _super);
        function StringType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        StringType.prototype.fromJsonValue = function (state, json, currentValue) {
            return _super.prototype.fromJsonValue.call(this, state, json, currentValue);
        };
        return StringType;
    }(BasicType))('String', String);
    BasicType.DateTime = new /** @class */ (function (_super) {
        __extends(DateTimeType, _super);
        function DateTimeType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        DateTimeType.prototype.toJsonValue = function (state, value) {
            return dateToJson(value);
        };
        DateTimeType.prototype.fromJsonValue = function (state, json, currentValue) {
            return jsonToDate(json, currentValue);
        };
        return DateTimeType;
    }(BasicType))('DateTime', Date);
    BasicType.Date = new /** @class */ (function (_super) {
        __extends(DateType, _super);
        function DateType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        DateType.prototype.toJsonValue = function (state, value) {
            var json = dateToJson(value);
            return json ? json.substring(0, json.indexOf('T')) : null;
        };
        DateType.prototype.fromJsonValue = function (state, json, currentValue) {
            return jsonToDate(json, currentValue);
        };
        return DateType;
    }(BasicType))('Date', Date);
    BasicType.Time = new /** @class */ (function (_super) {
        __extends(TimeType, _super);
        function TimeType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TimeType.prototype.toJsonValue = function (state, value) {
            var json = dateToJson(value);
            return json ? json.substring(json.indexOf('T') + 1) : null;
        };
        TimeType.prototype.fromJsonValue = function (state, json, currentValue) {
            return typeof json === 'string' ? jsonToDate("1970-01-01T" + json, currentValue) : null;
        };
        return TimeType;
    }(BasicType))('Time', Date);
    BasicType.File = new /** @class */ (function (_super) {
        __extends(FileType, _super);
        function FileType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FileType.prototype.toJsonValue = function (state, value) {
            return value instanceof binding_1.File ? value.id : null;
        };
        FileType.prototype.fromJsonValue = function (state, json, currentValue) {
            if (!json) {
                return null;
            }
            if (currentValue && currentValue.id === json) {
                return currentValue;
            }
            if (state.db) {
                return new state.db.File(json);
            }
            return null;
        };
        return FileType;
    }(BasicType))('File', binding_1.File);
    BasicType.GeoPoint = new /** @class */ (function (_super) {
        __extends(GeoPointType, _super);
        function GeoPointType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        GeoPointType.prototype.toJsonValue = function (state, value) {
            return value instanceof GeoPoint_1.GeoPoint ? value.toJSON() : null;
        };
        GeoPointType.prototype.fromJsonValue = function (state, json) {
            return json ? new GeoPoint_1.GeoPoint(json) : null;
        };
        return GeoPointType;
    }(BasicType))('GeoPoint', GeoPoint_1.GeoPoint);
    BasicType.JsonArray = new /** @class */ (function (_super) {
        __extends(JsonArrayType, _super);
        function JsonArrayType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        JsonArrayType.prototype.toJsonValue = function (state, value) {
            return Array.isArray(value) ? value : null;
        };
        JsonArrayType.prototype.fromJsonValue = function (state, json) {
            return Array.isArray(json) ? json : null;
        };
        return JsonArrayType;
    }(BasicType))('JsonArray', Array);
    BasicType.JsonObject = new /** @class */ (function (_super) {
        __extends(JsonObjectType, _super);
        function JsonObjectType() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        JsonObjectType.prototype.fromJsonValue = function (state, json, currentValue) {
            return _super.prototype.fromJsonValue.call(this, state, json, currentValue);
        };
        JsonObjectType.prototype.toJsonValue = function (state, value) {
            if (value && value.constructor === Object) {
                return value;
            }
            return null;
        };
        return JsonObjectType;
    }(BasicType))('JsonObject', Object);
    return BasicType;
}(Type_1.Type));
exports.BasicType = BasicType;
