// baqend/commonjs/lib/metamodel/EntityType.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityType = void 0;
// eslint-disable-next-line max-classes-per-file
var binding_1 = require("../binding");
var ManagedType_1 = require("./ManagedType");
var Type_1 = require("./Type");
var BasicType_1 = require("./BasicType");
var SingularAttribute_1 = require("./SingularAttribute");
var PluralAttribute_1 = require("./PluralAttribute");
var Attribute_1 = require("./Attribute");
var intersection_1 = require("../intersection");
var EntityType = /** @class */ (function (_super) {
    __extends(EntityType, _super);
    /**
     * @param ref
     * @param superType
     * @param typeConstructor
     */
    function EntityType(ref, superType, typeConstructor) {
        var _this = _super.call(this, ref, typeConstructor) || this;
        _this.declaredId = null;
        _this.declaredVersion = null;
        _this.declaredAcl = null;
        _this.loadPermission = new intersection_1.Permission();
        _this.updatePermission = new intersection_1.Permission();
        _this.deletePermission = new intersection_1.Permission();
        _this.queryPermission = new intersection_1.Permission();
        _this.schemaSubclassPermission = new intersection_1.Permission();
        _this.insertPermission = new intersection_1.Permission();
        _this.superType = superType;
        return _this;
    }
    Object.defineProperty(EntityType.prototype, "persistenceType", {
        /**
         * @inheritDoc
         */
        get: function () {
            return Type_1.PersistenceType.ENTITY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EntityType.prototype, "id", {
        get: function () {
            return this.declaredId || this.superType.id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EntityType.prototype, "version", {
        get: function () {
            return this.declaredVersion || this.superType.version;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(EntityType.prototype, "acl", {
        get: function () {
            return this.declaredAcl || this.superType.acl;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @inheritDoc
     */
    EntityType.prototype.createProxyClass = function () {
        var typeConstructor = this.superType.typeConstructor;
        if (typeConstructor === Object) {
            switch (this.name) {
                case 'User':
                    typeConstructor = binding_1.User;
                    break;
                case 'Role':
                    typeConstructor = binding_1.Role;
                    break;
                default:
                    typeConstructor = binding_1.Entity;
                    break;
            }
        }
        return this.enhancer.createProxy(typeConstructor);
    };
    /**
     * Gets all on this class referencing attributes
     *
     * @param db The instances will be found by this EntityManager
     * @param [options] Some options to pass
     * @param [options.classes] An array of class names to filter for, null for no filter
     * @return A map from every referencing class to a set of its referencing attribute names
     */
    EntityType.prototype.getReferencing = function (db, options) {
        var opts = __assign({}, options);
        var entities = db.metamodel.entities;
        var referencing = new Map();
        var names = Object.keys(entities);
        for (var i = 0, len = names.length; i < len; i += 1) {
            var name_1 = names[i];
            // Skip class if not in class filter
            if (!opts.classes || opts.classes.indexOf(name_1) !== -1) {
                var entity = entities[name_1];
                var iter = entity.attributes();
                for (var el = iter.next(); !el.done; el = iter.next()) {
                    var attr = el.value;
                    // Filter only referencing singular and collection attributes
                    if ((attr instanceof SingularAttribute_1.SingularAttribute && attr.type === this)
                        || (attr instanceof PluralAttribute_1.PluralAttribute && attr.elementType === this)) {
                        var typeReferences = referencing.get(attr.declaringType) || new Set();
                        typeReferences.add(attr.name);
                        referencing.set(attr.declaringType, typeReferences);
                    }
                }
            }
        }
        return referencing;
    };
    /**
     * @inheritDoc
     */
    EntityType.prototype.createObjectFactory = function (db) {
        switch (this.name) {
            case 'User':
                return binding_1.UserFactory.create(this, db);
            case 'Device':
                return binding_1.DeviceFactory.create(this, db);
            default:
                return binding_1.EntityFactory.create(this, db);
        }
    };
    /**
     * @param state The root object state, can be <code>null</code> if a currentObject is provided
     * @param jsonObject The json data to merge
     * @param currentObject The object where the jsonObject will be merged into, if the current object is null,
     * a new instance will be created
     * @param options The options used to apply the json
     * @param [options.persisting=false] indicates if the current state will be persisted.
     * Used to update the internal change tracking state of collections and mark the object persistent or dirty afterwards
     * @param [options.onlyMetadata=false] Indicates if only the metadata should be updated
     * @return The merged entity instance
     */
    EntityType.prototype.fromJsonValue = function (state, jsonObject, currentObject, options) {
        var _a, _b;
        // handle references
        if (typeof jsonObject === 'string') {
            return ((_a = state.db) === null || _a === void 0 ? void 0 : _a.getReference(jsonObject)) || null;
        }
        if (!jsonObject || typeof jsonObject !== 'object') {
            return null;
        }
        var json = jsonObject;
        var opt = __assign({ persisting: false, onlyMetadata: false }, options);
        var obj;
        if (currentObject) {
            var currentObjectState = intersection_1.Metadata.get(currentObject);
            // merge state into the current object if:
            // 1. The provided json does not contains an id and we have an already created object for it
            // 2. The object was created without an id and was later fetched from the server (e.g. User/Role)
            // 3. The provided json has the same id as the current object, they can differ on embedded json for a reference
            if (!json.id || !currentObjectState.id || json.id === currentObjectState.id) {
                obj = currentObject;
            }
        }
        if (!obj) {
            obj = (_b = state.db) === null || _b === void 0 ? void 0 : _b.getReference(this.typeConstructor, json.id);
        }
        if (!obj) {
            return null;
        }
        var objectState = intersection_1.Metadata.get(obj);
        // deserialize our properties
        objectState.enable(false);
        _super.prototype.fromJsonValue.call(this, objectState, json, obj, opt);
        objectState.enable(true);
        if (opt.persisting) {
            objectState.setPersistent();
        }
        else if (!opt.onlyMetadata) {
            objectState.setDirty();
        }
        return obj;
    };
    /**
     * Converts the given object to json
     * @param state The root object state
     * @param object The object to convert
     * @param [options=false] to json options by default excludes the metadata
     * @param [options.excludeMetadata=false] Excludes the metadata form the serialized json
     * @param [options.depth=0] Includes up to depth referenced objects into the serialized json
     * @param [options.persisting=false] indicates if the current state will be persisted.
     *  Used to update the internal change tracking state of collections and mark the object persistent if its true
     * @return JSON-Object
     */
    EntityType.prototype.toJsonValue = function (state, object, options) {
        var _a = options || {}, _b = _a.depth, depth = _b === void 0 ? 0 : _b, _c = _a.persisting, persisting = _c === void 0 ? false : _c;
        var isInDepth = depth === true || depth > -1;
        // check if object is already loaded in state
        var objectState = object && intersection_1.Metadata.get(object);
        if (isInDepth && objectState && objectState.isAvailable) {
            // serialize our properties
            objectState.enable(false);
            var json = _super.prototype.toJsonValue.call(this, objectState, object, __assign(__assign({}, options), { persisting: persisting, depth: typeof depth === 'boolean' ? depth : depth - 1 }));
            objectState.enable(true);
            return json;
        }
        if (state.db && object instanceof this.typeConstructor) {
            object.attach(state.db);
            return object.id;
        }
        return null;
    };
    EntityType.prototype.toString = function () {
        return "EntityType(" + this.ref + ")";
    };
    EntityType.prototype.toJSON = function () {
        var _a = _super.prototype.toJSON.call(this), acl = _a.acl, json = __rest(_a, ["acl"]);
        return __assign(__assign({}, json), { acl: __assign(__assign({}, acl), { schemaSubclass: this.schemaSubclassPermission.toJSON(), load: this.loadPermission.toJSON(), insert: this.insertPermission.toJSON(), update: this.updatePermission.toJSON(), delete: this.deletePermission.toJSON(), query: this.queryPermission.toJSON() }) });
    };
    EntityType.Object = /** @class */ (function (_super) {
        __extends(ObjectType, _super);
        function ObjectType() {
            var _this = _super.call(this, EntityType.Object.ref, null, Object) || this;
            _this.declaredId = new /** @class */ (function (_super) {
                __extends(class_1, _super);
                function class_1() {
                    return _super.call(this, 'id', BasicType_1.BasicType.String, true) || this;
                }
                class_1.prototype.getJsonValue = function (state) {
                    return state.id || undefined;
                };
                class_1.prototype.setJsonValue = function (state, object, jsonValue) {
                    if (!state.id) {
                        // eslint-disable-next-line no-param-reassign
                        state.id = jsonValue;
                    }
                };
                return class_1;
            }(SingularAttribute_1.SingularAttribute))();
            _this.declaredId.init(_this, 0);
            _this.declaredId.isId = true;
            _this.declaredVersion = new /** @class */ (function (_super) {
                __extends(class_2, _super);
                function class_2() {
                    return _super.call(this, 'version', BasicType_1.BasicType.Integer, true) || this;
                }
                class_2.prototype.getJsonValue = function (state) {
                    return state.version || undefined;
                };
                class_2.prototype.setJsonValue = function (state, object, jsonValue) {
                    if (jsonValue) {
                        // eslint-disable-next-line no-param-reassign
                        state.version = jsonValue;
                    }
                };
                return class_2;
            }(SingularAttribute_1.SingularAttribute))();
            _this.declaredVersion.init(_this, 1);
            _this.declaredVersion.isVersion = true;
            _this.declaredAcl = new /** @class */ (function (_super) {
                __extends(class_3, _super);
                function class_3() {
                    return _super.call(this, 'acl', BasicType_1.BasicType.JsonObject, true) || this;
                }
                class_3.prototype.getJsonValue = function (state, object, options) {
                    var persisted = Attribute_1.Attribute.attachState(object, {});
                    var persistedAcl = persisted.acl || {};
                    var acl = state.acl.toJSON();
                    var unchanged = Object.keys(acl).every(function (permission) {
                        var oldPermission = (persistedAcl[permission] || {});
                        var newPermission = acl[permission];
                        var newKeys = Object.keys(newPermission);
                        var oldKeys = Object.keys(oldPermission);
                        return newKeys.length === oldKeys.length
                            && newKeys.every(function (ref) { return oldPermission[ref] === newPermission[ref]; });
                    });
                    if (!unchanged) {
                        state.setDirty();
                    }
                    if (options.persisting) {
                        persisted.acl = acl;
                    }
                    return acl;
                };
                class_3.prototype.setJsonValue = function (state, object, jsonValue, options) {
                    var acl = (jsonValue || {});
                    if (options.persisting) {
                        var persistedState = Attribute_1.Attribute.attachState(object, {});
                        persistedState.acl = acl;
                    }
                    state.acl.fromJSON(acl);
                };
                return class_3;
            }(SingularAttribute_1.SingularAttribute))();
            _this.declaredAcl.init(_this, 2);
            _this.declaredAcl.isAcl = true;
            _this.declaredAttributes = [_this.declaredId, _this.declaredVersion, _this.declaredAcl];
            return _this;
        }
        Object.defineProperty(ObjectType, "ref", {
            get: function () {
                return '/db/Object';
            },
            enumerable: false,
            configurable: true
        });
        ObjectType.prototype.createProxyClass = function () {
            return _super.prototype.createProxyClass.call(this);
        };
        ObjectType.prototype.fromJsonValue = function (state, jsonObject, currentObject, options) {
            return _super.prototype.fromJsonValue.call(this, state, jsonObject, currentObject, options);
        };
        ObjectType.prototype.createObjectFactory = function () {
            throw new Error("Objects can't be directly created and persisted");
        };
        return ObjectType;
    }(EntityType));
    return EntityType;
}(ManagedType_1.ManagedType));
exports.EntityType = EntityType;
