// baqend/commonjs/lib/metamodel/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PersistentAttributeType = exports.Attribute = exports.BasicType = exports.CollectionAttribute = exports.EmbeddableType = exports.EntityType = exports.ListAttribute = exports.ManagedType = exports.MapAttribute = exports.ModelBuilder = exports.CollectionType = exports.PluralAttribute = exports.SetAttribute = exports.SingularAttribute = exports.PersistenceType = exports.Type = exports.DbIndex = exports.Metamodel = void 0;
var Metamodel_1 = require("./Metamodel");
Object.defineProperty(exports, "Metamodel", { enumerable: true, get: function () { return Metamodel_1.Metamodel; } });
var DbIndex_1 = require("./DbIndex");
Object.defineProperty(exports, "DbIndex", { enumerable: true, get: function () { return DbIndex_1.DbIndex; } });
var Type_1 = require("./Type");
Object.defineProperty(exports, "Type", { enumerable: true, get: function () { return Type_1.Type; } });
Object.defineProperty(exports, "PersistenceType", { enumerable: true, get: function () { return Type_1.PersistenceType; } });
var SingularAttribute_1 = require("./SingularAttribute");
Object.defineProperty(exports, "SingularAttribute", { enumerable: true, get: function () { return SingularAttribute_1.SingularAttribute; } });
var SetAttribute_1 = require("./SetAttribute");
Object.defineProperty(exports, "SetAttribute", { enumerable: true, get: function () { return SetAttribute_1.SetAttribute; } });
var PluralAttribute_1 = require("./PluralAttribute");
Object.defineProperty(exports, "PluralAttribute", { enumerable: true, get: function () { return PluralAttribute_1.PluralAttribute; } });
Object.defineProperty(exports, "CollectionType", { enumerable: true, get: function () { return PluralAttribute_1.CollectionType; } });
var ModelBuilder_1 = require("./ModelBuilder");
Object.defineProperty(exports, "ModelBuilder", { enumerable: true, get: function () { return ModelBuilder_1.ModelBuilder; } });
var MapAttribute_1 = require("./MapAttribute");
Object.defineProperty(exports, "MapAttribute", { enumerable: true, get: function () { return MapAttribute_1.MapAttribute; } });
var ManagedType_1 = require("./ManagedType");
Object.defineProperty(exports, "ManagedType", { enumerable: true, get: function () { return ManagedType_1.ManagedType; } });
var ListAttribute_1 = require("./ListAttribute");
Object.defineProperty(exports, "ListAttribute", { enumerable: true, get: function () { return ListAttribute_1.ListAttribute; } });
var EntityType_1 = require("./EntityType");
Object.defineProperty(exports, "EntityType", { enumerable: true, get: function () { return EntityType_1.EntityType; } });
var EmbeddableType_1 = require("./EmbeddableType");
Object.defineProperty(exports, "EmbeddableType", { enumerable: true, get: function () { return EmbeddableType_1.EmbeddableType; } });
var CollectionAttribute_1 = require("./CollectionAttribute");
Object.defineProperty(exports, "CollectionAttribute", { enumerable: true, get: function () { return CollectionAttribute_1.CollectionAttribute; } });
var BasicType_1 = require("./BasicType");
Object.defineProperty(exports, "BasicType", { enumerable: true, get: function () { return BasicType_1.BasicType; } });
var Attribute_1 = require("./Attribute");
Object.defineProperty(exports, "Attribute", { enumerable: true, get: function () { return Attribute_1.Attribute; } });
Object.defineProperty(exports, "PersistentAttributeType", { enumerable: true, get: function () { return Attribute_1.PersistentAttributeType; } });
