// baqend/commonjs/lib/connector/Message.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuthMessage = exports.Message = exports.appendQueryParams = exports.StatusCode = void 0;
var error_1 = require("../error");
/**
 * Checks whether the user uses a browser which does support revalidation.
 */
// @ts-ignore
var REVALIDATION_SUPPORTED = typeof navigator === 'undefined' || (typeof chrome !== 'undefined' && /google/i.test(navigator.vendor)) || (/cros i686/i.test(navigator.platform));
exports.StatusCode = {
    NOT_MODIFIED: 304,
    BAD_CREDENTIALS: 460,
    BUCKET_NOT_FOUND: 461,
    INVALID_PERMISSION_MODIFICATION: 462,
    INVALID_TYPE_VALUE: 463,
    OBJECT_NOT_FOUND: 404,
    OBJECT_OUT_OF_DATE: 412,
    PERMISSION_DENIED: 466,
    QUERY_DISPOSED: 467,
    QUERY_NOT_SUPPORTED: 468,
    SCHEMA_NOT_COMPATIBLE: 469,
    SCHEMA_STILL_EXISTS: 470,
    SYNTAX_ERROR: 471,
    TRANSACTION_INACTIVE: 472,
    TYPE_ALREADY_EXISTS: 473,
    TYPE_STILL_REFERENCED: 474,
    SCRIPT_ABORTION: 475,
};
/**
 * Appends the given query parameters to the url
 * @param url - on which the parameters will be appended
 * @param queryParams - The Query parameters which should be appended
 * @return The URL with the appended parameters
 */
function appendQueryParams(url, queryParams) {
    var queryString = typeof queryParams === 'string' ? queryParams : Object.entries(queryParams)
        .filter(function (_a) {
        var value = _a[1];
        return value !== undefined;
    })
        .map(function (_a) {
        var key = _a[0], value = _a[1];
        return key + "=" + encodeURIComponent(value);
    })
        .join('&');
    if (!queryString) {
        return url;
    }
    var sep = url.indexOf('?') >= 0 ? '&' : '?';
    return url + sep + queryString;
}
exports.appendQueryParams = appendQueryParams;
var Message = /** @class */ (function () {
    /**
     * @param args The path arguments
     */
    function Message() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.withCredentials = false;
        this.progressCallback = null;
        this._tokenStorage = null;
        this._responseType = null;
        var index = 0;
        var path = this.spec.path[0];
        var len = this.spec.path.length;
        for (var i = 1; i < len; i += 1) {
            if (this.spec.dynamic && len - 1 === i) {
                path += args[index].split('/').map(encodeURIComponent).join('/');
            }
            else {
                path += encodeURIComponent(args[index]) + this.spec.path[i];
            }
            index += 1;
        }
        var queryParams = {};
        for (var i = 0; i < this.spec.query.length; i += 1) {
            var arg = args[index];
            index += 1;
            if (arg !== undefined && arg !== null) {
                queryParams[this.spec.query[i]] = arg;
            }
        }
        this.request = {
            method: this.spec.method,
            path: appendQueryParams(path, queryParams),
            entity: null,
            headers: {},
        };
        if (args[index]) {
            this.entity(args[index], 'json');
        }
        this.responseType('json');
    }
    Object.defineProperty(Message.prototype, "spec", {
        /**
         * Returns the specification of this message
         */
        get: function () { return null; },
        enumerable: false,
        configurable: true
    });
    /**
     * Creates a new message class with the given message specification
     * @return A created message object for the specification
     */
    Message.create = function (specification) {
        var parts = specification.path.split('?');
        var path = parts[0].split(/[:*]\w*/);
        var query = [];
        if (parts[1]) {
            parts[1].split('&').forEach(function (arg) {
                var part = arg.split('=');
                query.push(part[0]);
            });
        }
        var spec = {
            path: path,
            query: query,
            status: specification.status,
            method: specification.method,
            dynamic: specification.path.indexOf('*') !== -1,
        };
        return /** @class */ (function (_super) {
            __extends(class_1, _super);
            function class_1() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            Object.defineProperty(class_1.prototype, "spec", {
                get: function () {
                    return spec;
                },
                enumerable: false,
                configurable: true
            });
            return class_1;
        }(Message));
    };
    Object.defineProperty(Message.prototype, "isBinary", {
        get: function () {
            return (this.request.type && this.request.type in Message.BINARY) || this._responseType in Message.BINARY;
        },
        enumerable: false,
        configurable: true
    });
    Message.prototype.tokenStorage = function (value) {
        if (value === undefined) {
            return this._tokenStorage;
        }
        this._tokenStorage = value;
        return this;
    };
    Message.prototype.path = function (path) {
        if (path !== undefined) {
            var queryIndex = this.request.path.indexOf('?') + 1;
            this.request.path = path + (queryIndex > 0 ? (path.indexOf('?') > -1 ? '&' : '?') + this.request.path.substring(queryIndex) : '');
            return this;
        }
        return this.request.path;
    };
    Message.prototype.header = function (name, value) {
        if (value === null) {
            delete this.request.headers[name];
            return this;
        }
        if (value !== undefined) {
            this.request.headers[name] = value;
            return this;
        }
        return this.request.headers[name];
    };
    /**
     * Sets the entity type
     * @param data - The data to send
     * @param type - the type of the data one of 'json'|'text'|'blob'|'arraybuffer'
     * defaults detect the type based on the body data
     * @return This message object
     */
    Message.prototype.entity = function (data, type) {
        var requestType = type;
        if (!requestType) {
            if (typeof data === 'string') {
                if (/^data:(.+?)(;base64)?,.*$/.test(data)) {
                    requestType = 'data-url';
                }
                else {
                    requestType = 'text';
                }
            }
            else if (typeof Blob !== 'undefined' && data instanceof Blob) {
                requestType = 'blob';
            }
            else if (typeof Buffer !== 'undefined' && data instanceof Buffer) {
                requestType = 'buffer';
            }
            else if (typeof ArrayBuffer !== 'undefined' && data instanceof ArrayBuffer) {
                requestType = 'arraybuffer';
            }
            else if (typeof FormData !== 'undefined' && data instanceof FormData) {
                requestType = 'form';
            }
            else {
                requestType = 'json';
            }
        }
        this.request.type = requestType;
        this.request.entity = data;
        return this;
    };
    Message.prototype.mimeType = function (mimeType) {
        return this.header('content-type', mimeType);
    };
    Message.prototype.contentLength = function (contentLength) {
        if (contentLength !== undefined) {
            return this.header('content-length', "" + contentLength);
        }
        return Number(this.header('content-length'));
    };
    Message.prototype.ifMatch = function (eTag) {
        return this.header('If-Match', this.formatETag(eTag));
    };
    Message.prototype.ifNoneMatch = function (eTag) {
        return this.header('If-None-Match', this.formatETag(eTag));
    };
    Message.prototype.ifUnmodifiedSince = function (date) {
        // IE 10 returns UTC strings and not an RFC-1123 GMT date string
        return this.header('if-unmodified-since', date && date.toUTCString().replace('UTC', 'GMT'));
    };
    /**
     * Indicates that the request should not be served by a local cache
     * @return
     */
    Message.prototype.noCache = function () {
        if (!REVALIDATION_SUPPORTED) {
            this.ifMatch('') // is needed for firefox or safari (but forbidden for chrome)
                .ifNoneMatch('-'); // is needed for edge and ie (but forbidden for chrome)
        }
        return this.cacheControl('max-age=0, no-cache');
    };
    Message.prototype.cacheControl = function (value) {
        return this.header('cache-control', value);
    };
    Message.prototype.acl = function (acl) {
        return this.header('baqend-acl', acl && JSON.stringify(acl));
    };
    Message.prototype.customHeaders = function (customHeaders) {
        return this.header('baqend-custom-headers', customHeaders && JSON.stringify(customHeaders));
    };
    Message.prototype.accept = function (accept) {
        return this.header('accept', accept);
    };
    Message.prototype.responseType = function (type) {
        if (type !== undefined) {
            this._responseType = type;
            return this;
        }
        return this._responseType;
    };
    Message.prototype.progress = function (callback) {
        if (callback !== undefined) {
            this.progressCallback = callback;
            return this;
        }
        return this.progressCallback;
    };
    /**
     * Adds the given string to the request path
     *
     * If the parameter is an object, it will be serialized as a query string.
     *
     * @param query which will added to the request path
     * @return
     */
    Message.prototype.addQueryString = function (query) {
        this.request.path = appendQueryParams(this.request.path, query);
        return this;
    };
    Message.prototype.formatETag = function (eTag) {
        if (eTag === null || eTag === undefined || eTag === '*') {
            return eTag;
        }
        var tag = "" + eTag;
        if (tag.indexOf('"') === -1) {
            tag = "\"" + tag + "\"";
        }
        return tag;
    };
    /**
     * Handle the receive
     * @param response The received response headers and data
     * @return
     */
    Message.prototype.doReceive = function (response) {
        if (this.spec.status.indexOf(response.status) === -1) {
            throw new error_1.CommunicationError(this, response);
        }
    };
    Message.StatusCode = exports.StatusCode;
    Message.BINARY = {
        blob: true,
        buffer: true,
        stream: true,
        arraybuffer: true,
        'data-url': true,
        base64: true,
    };
    return Message;
}());
exports.Message = Message;
var OAuthMessage = /** @class */ (function (_super) {
    __extends(OAuthMessage, _super);
    function OAuthMessage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(OAuthMessage.prototype, "spec", {
        get: function () {
            return {
                method: 'OAUTH',
                dynamic: false,
                path: [''],
                query: [],
                status: [200],
            };
        },
        enumerable: false,
        configurable: true
    });
    return OAuthMessage;
}(Message));
exports.OAuthMessage = OAuthMessage;
