// baqend/cli/typings.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.typings = void 0;
/* eslint-disable no-continue */
const os_1 = __importDefault(require("os"));
const baqend_1 = require("baqend");
const helper_1 = require("./helper");
const { CollectionType } = baqend_1.metamodel.PluralAttribute;
const tsTypeMapping = {
    Boolean: 'boolean',
    Double: 'number',
    Integer: 'number',
    String: 'string',
    DateTime: 'Date',
    Time: 'Date',
    Date: 'Date',
    GeoPoint: 'GeoPoint',
    JsonArray: '[]',
    JsonObject: '{}',
    File: 'binding.File',
};
const nativeTypes = ['User', 'Role', 'Device'];
const push = Function.prototype.apply.bind(Array.prototype.push);
function typings(args) {
    if (!args.app) {
        return Promise.reject(new Error('Please specify the app parameter.'));
    }
    return createTypings(args.app, args.dest);
}
exports.typings = typings;
function createTypings(app, dest) {
    const factory = new baqend_1.EntityManagerFactory({ host: app });
    return factory.ready().then(() => {
        const types = typingsFromMetamodel(factory.metamodel);
        return helper_1.ensureDir(dest)
            .then(() => helper_1.writeFile(`${dest}/baqend-model.d.ts`, types.join(os_1.default.EOL)))
            .then(() => console.log(`Typings successfully saved to: ${dest}/baqend-model.d.ts`));
    });
}
function typingsFromMetamodel(metamodel) {
    const typing = [];
    const namespaces = {};
    // import all native types, so they can be easily used in definitions
    typing.push('import {binding, GeoPoint} from "baqend";');
    typing.push('');
    const module = [];
    module.push('declare module "baqend" {');
    module.push('');
    module.push('  interface baqend {');
    const model = [];
    model.push('  namespace model {');
    for (const key of Object.keys(metamodel.entities)) {
        const entity = metamodel.entities[key];
        if (entity.name === 'Object') {
            continue;
        }
        if (nativeTypes.includes(entity.name)) {
            continue;
        }
        if (helper_1.isNativeClassNamespace(entity.name)) {
            continue;
        }
        // register only user defined types
        if (entity.name.indexOf('.') !== -1) {
            const [namespace, entityName] = entity.name.split('.');
            module.push(`    ["${entity.name}"]: binding.EntityFactory<model.${entity.name}>;`);
            if (!namespaces[namespace])
                namespaces[namespace] = [];
            push(namespaces[namespace], typingsFromSchema(entityName, entity, 'Entity'));
        }
        else {
            module.push(`    ${entity.name}: binding.EntityFactory<model.${entity.name}>;`);
            push(model, typingsFromSchema(entity.name, entity, 'Entity'));
        }
        model.push('');
    }
    for (const key of Object.keys(metamodel.embeddables)) {
        const embeddable = metamodel.embeddables[key];
        if (embeddable.name.indexOf('.') !== -1)
            continue;
        module.push(`    ${embeddable.name}: binding.ManagedFactory<model.${embeddable.name}>;`);
        push(model, typingsFromSchema(embeddable.name, embeddable, 'Managed'));
        model.push('');
    }
    module.push('  }');
    module.push('');
    for (const key of Object.keys(namespaces)) {
        model.push(`  namespace ${key} {`);
        push(model, namespaces[key]);
        model.push('  }');
    }
    model.push('  }');
    push(module, model);
    module.push('}');
    push(typing, module);
    return typing;
}
function typingsFromSchema(typeName, entity, type) {
    const typing = [];
    typing.push(`    interface ${typeName} extends binding.${type} {`);
    for (const attribute of entity.declaredAttributes) {
        if (!attribute.isMetadata) {
            if (attribute.isCollection) {
                switch (attribute.collectionType) {
                    case CollectionType.LIST:
                        typing.push(`      ${attribute.name}: Array<${typingsFromType(attribute.elementType)}>;`);
                        break;
                    case CollectionType.MAP: {
                        const mapAttr = attribute;
                        typing.push(`      ${attribute.name}: Map<${typingsFromType(mapAttr.keyType)}, ${typingsFromType(mapAttr.elementType)}>;`);
                        break;
                    }
                    case CollectionType.SET:
                        typing.push(`      ${attribute.name}: Set<${typingsFromType(attribute.elementType)}>;`);
                        break;
                    default:
                        break;
                }
            }
            else {
                typing.push(`      ${attribute.name}: ${typingsFromType(attribute.type)};`);
            }
        }
    }
    typing.push('    }');
    return typing;
}
function typingsFromType(type) {
    if (type.isBasic) {
        return tsTypeMapping[type.name];
    }
    return type.name.split('.').pop();
}
/**
 import {binding} from "baqend";

 declare module "baqend" {

  interface baqend {
    Campaign:binding.EntityFactory<Campaign>;
    Profile:binding.EntityFactory<Profile>;
  }

  export interface Campaign extends binding.Entity {

  }

  export interface Profile extends binding.Entity {

  }

}
* */
