// baqend/cli/account.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.whoami = exports.listApps = exports.openDashboard = exports.openApp = exports.persistLogin = exports.logout = exports.login = exports.register = void 0;
const open_1 = __importDefault(require("open"));
const crypto_1 = __importDefault(require("crypto"));
const http_1 = require("http");
const os_1 = __importDefault(require("os"));
const baqend_1 = require("baqend");
const helper = __importStar(require("./helper"));
const helper_1 = require("./helper");
const { TokenStorage } = baqend_1.intersection;
const { UserFactory } = baqend_1.binding;
const fileName = `${os_1.default.homedir()}/.baqend`;
const algorithm = 'aes-256-ctr';
const PROFILE_DEFAULT_KEY = 'N2Ki=za[8iy4ff4jYn/3,y;';
const bbqHost = 'bbq';
function getAppInfo(args) {
    const isCustomHost = args.app && /^https?:\/\//.test(args.app);
    return {
        isCustomHost: !!isCustomHost,
        host: isCustomHost ? args.app : bbqHost,
        app: isCustomHost ? null : args.app,
    };
}
function getArgsCredentials(args) {
    if (args.username && args.password) {
        return {
            username: args.username,
            password: args.password,
        };
    }
    if (args.token) {
        return { token: args.token };
    }
    return null;
}
function getEnvCredentials() {
    const token = process.env.BAQEND_TOKEN || process.env.BAT;
    if (token) {
        return { token };
    }
    return null;
}
function getProfileCredentials(appInfo) {
    return readProfileFile().then((json) => {
        const credentials = json[appInfo.host];
        if (!credentials) {
            return null;
        }
        if ('password' in credentials) {
            console.log('Storing username/password in the baqend profile will not be supported in future version.');
            console.log('Logout and login again to fix this issue.');
            credentials.password = decrypt(credentials.password);
        }
        return credentials;
    }).catch(() => null);
}
function getLocalCredentials(appInfo) {
    if (appInfo.isCustomHost) {
        return { username: 'root', password: 'root' };
    }
    return null;
}
function getInputCredentials(appInfo, authProvider, showLoginInfo) {
    if (!process.stdout.isTTY) {
        return Promise.reject(new Error('Can\'t interactive login into baqend, no tty session was detected.'));
    }
    if (showLoginInfo) {
        console.log('Baqend Login is required. You can skip this step by saving the Login credentials with "baqend login or baqend sso"');
    }
    const options = ['password', 'google', 'facebook', 'github'];
    let result = Promise.resolve(String(options.indexOf(authProvider || 'password') + 1));
    if (!appInfo.isCustomHost && options.length > 1 && !authProvider) {
        console.log('Choose how you want to login:');
        options.forEach((provider, index) => {
            console.log(`${index + 1}. Login with ${provider}`);
        });
        result = helper_1.readInput(`Type 1-${options.length}: `);
    }
    return result.then((option) => {
        if (option === '1') {
            return readInputCredentials(appInfo);
        }
        const provider = options[Number(option) - 1];
        if (!provider) {
            throw new Error('No valid login option was chosen.');
        }
        return requestSSOCredentials(appInfo, provider);
    });
}
function requestSSOCredentials(appInfo, oAuthProvider, oAuthOptions = {}) {
    // TODO: current workaround until our server pass this ids to the client dynamically
    const clientIds = {
        facebook: '976707865723719',
        google: '586076830320-0el1jebupjvbcmqf95vfaqjq7gbs0bdh.apps.googleusercontent.com',
        gitHub: '1311e3415ab415fda705',
    };
    return appConnect(appInfo)
        .then((db) => {
        const host = '127.0.0.1';
        const port = 9876;
        const provider = oAuthProvider.toLowerCase();
        return Promise.all([
            oAuthHandler(host, port),
            db.loginWithOAuth(oAuthProvider, Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, UserFactory.DefaultOptions[provider] || {}), (clientIds[provider] && { clientId: clientIds[provider] })), { open: open_1.default }), oAuthOptions), { redirect: `http://${host}:${port}` })),
        ]).then(([credentials, windowProcess]) => {
            windowProcess.kill('SIGHUP'); // seems not working on every platform
            return credentials;
        });
    });
}
function oAuthHandler(host, port) {
    return new Promise((resolve, reject) => {
        const server = http_1.createServer((req, res) => {
            const url = new URL(req.url, `http://${host}:${port}`);
            let done = false;
            if (url.searchParams.has('errorMessage')) {
                reject(new Error(url.searchParams.get('errorMessage')));
                done = true;
            }
            else if (url.searchParams.has('token')) {
                resolve({ token: url.searchParams.get('token') });
                done = true;
            }
            if (done) {
                res.writeHead(200, { 'Content-Type': 'text/html' });
                res.end('<!DOCTYPE html><html><head></head><body><h1>Continue within the CLI!</h1></a></body></html>');
                server.close();
            }
            else {
                res.writeHead(404);
                res.end();
            }
        });
        server.on('error', (err) => {
            reject(err);
        });
        server.listen(port, host);
    });
}
function getCredentials(appInfo, args) {
    let providers = Promise.resolve(null)
        .then((credentials) => credentials || getArgsCredentials(args))
        .then((credentials) => credentials || getEnvCredentials())
        .then((credentials) => credentials || getProfileCredentials(appInfo))
        .then((credentials) => credentials || getLocalCredentials(appInfo));
    if (!args.skipInput && process.stdout.isTTY) {
        providers = providers
            .then((credentials) => credentials || getInputCredentials(appInfo, args.auth, true));
    }
    return providers;
}
function register(args) {
    const appInfo = getAppInfo(args);
    return getInputCredentials(appInfo, args.auth)
        .then((credentials) => appConnect(appInfo)
        .then((db) => {
        if ('token' in credentials) {
            return db.User.loginWithToken(credentials.token).then(() => db);
        }
        return db.User.register(credentials.username, credentials.password).then(() => db);
    })
        .then((db) => Promise.all([
        getDefaultApp(db).then((name) => console.log(`Your app name is ${name}`)),
        saveCredentials(appInfo, { token: db.token }),
    ])));
}
exports.register = register;
function connect(args) {
    const appInfo = getAppInfo(args);
    return getCredentials(appInfo, args)
        .then((credentials) => {
        if (!credentials)
            throw new Error('Login information are missing. Login with baqend login or pass a baqend token via BAQEND_TOKEN environment variable');
        return appConnect(appInfo, credentials);
    });
}
function appConnect(appInfo, credentials) {
    // do not use the global token storage here, to prevent login collisions on the bbq app
    const factory = new baqend_1.EntityManagerFactory({ host: appInfo.host, secure: true, tokenStorageFactory: TokenStorage });
    return factory.createEntityManager(true).ready().then((db) => {
        if (!credentials)
            return db;
        if ('token' in credentials) {
            return db.User.loginWithToken(credentials.token)
                .then((me) => {
                if (me) {
                    return db;
                }
                throw new Error('Login with Baqend token failed.');
            });
        }
        return db.User.login(credentials.username, credentials.password).then(() => db);
    });
}
function login(args) {
    const appInfo = getAppInfo(args);
    return connect(args)
        .then((db) => {
        if (appInfo.isCustomHost) {
            return db;
        }
        if (appInfo.app) {
            return bbqAppLogin(db, appInfo.app);
        }
        return getDefaultApp(db).then((appName) => bbqAppLogin(db, appName));
    }).catch((e) => {
        // if the login failed try to directly login into the app
        if (appInfo.app && !appInfo.isCustomHost) {
            return login(Object.assign(Object.assign({}, args), { app: `https://${appInfo.app}.app.baqend.com/v1` }));
        }
        throw e;
    });
}
exports.login = login;
function bbqAppLogin(db, appName) {
    return db.modules.get('apps', { app: appName }).then((result) => {
        if (!result) {
            throw new Error(`App (${appName}) not found.`);
        }
        return appConnect({ host: result.name, isCustomHost: false, app: appName }, { token: result.token });
    });
}
function logout(args) {
    const appInfo = getAppInfo(args);
    return readProfileFile().then((json) => {
        // eslint-disable-next-line no-param-reassign
        delete json[appInfo.host];
        return writeProfileFile(json);
    });
}
exports.logout = logout;
function persistLogin(args) {
    const appInfo = getAppInfo(args);
    let credentials = getArgsCredentials(args);
    if (!credentials) {
        credentials = getInputCredentials(appInfo, args.auth, false);
    }
    return Promise.resolve(credentials)
        .then((creds) => appConnect(appInfo, creds)
        .then((db) => saveCredentials(appInfo, 'token' in creds ? creds : { token: db.token })))
        .then(() => console.log('You have successfully been logged in.'));
}
exports.persistLogin = persistLogin;
function openApp(app) {
    if (app) {
        return open_1.default(`https://${app}.app.baqend.com`);
    }
    return login({}).then((db) => {
        open_1.default(`https://${db.connection.host}`);
    });
}
exports.openApp = openApp;
function openDashboard(args) {
    return connect(args).then((db) => {
        open_1.default(`https://dashboard.baqend.com/login?token=${db.token}`);
    }).catch(() => {
        open_1.default('https://dashboard.baqend.com');
    });
}
exports.openDashboard = openDashboard;
function listApps(args) {
    return connect(args)
        .then((db) => getApps(db))
        .then((apps) => apps.forEach((app) => console.log(app)));
}
exports.listApps = listApps;
function whoami(args) {
    return connect(Object.assign({ skipInput: true }, args))
        .then((db) => console.log(db.User.me.username), () => console.log('You are not logged in.'));
}
exports.whoami = whoami;
function getApps(db) {
    return db.modules.get('apps').then((apps) => apps.map((app) => app.name));
}
function getDefaultApp(db) {
    return getApps(db).then((apps) => {
        if (apps.length === 1) {
            return apps[0];
        }
        throw new Error('Please add the name of your app as a parameter.');
    });
}
function readInputCredentials(appInfo) {
    return helper.readInput(appInfo.isCustomHost ? 'Username: ' : 'E-Mail: ')
        .then((username) => helper.readInput('Password: ', true).then((password) => {
        console.log();
        return { username, password };
    }));
}
function decrypt(input) {
    // This is legacy and we will remove support for the username / password storage in the profile file
    // eslint-disable-next-line node/no-deprecated-api
    const decipher = crypto_1.default.createDecipher(algorithm, PROFILE_DEFAULT_KEY);
    let decrypted = decipher.update(input, 'base64', 'utf8');
    decrypted += decipher.final('utf8');
    return decrypted;
}
function writeProfileFile(json) {
    return helper.writeFile(fileName, JSON.stringify(json)).catch((e) => {
        console.warn('Baqend Profile file can\'t be written', e);
        throw e;
    });
}
function readProfileFile() {
    return helper_1.isFile(fileName).then((exists) => {
        if (!exists) {
            return {};
        }
        return helper_1.readFile(fileName, 'utf-8').then((data) => (data ? JSON.parse(data) : {}));
    });
}
function saveCredentials(appInfo, credentials) {
    return readProfileFile().then((json) => {
        // eslint-disable-next-line no-param-reassign
        json[appInfo.host] = Object.assign(Object.assign({}, json[appInfo.host]), credentials);
        return writeProfileFile(json);
    });
}
