// baqend/cli/copy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.copy = void 0;
const fs = __importStar(require("fs"));
const account = __importStar(require("./account"));
const helper_1 = require("./helper");
function splitArg(arg) {
    const index = arg.lastIndexOf(':');
    // Has app and path part?
    if (index >= 0) {
        const app = arg.substring(0, index);
        const path = arg.substring(index + 1);
        // Add www bucket prefix if path is relative
        const absolutePath = path.startsWith('/') ? path : `/www/${path}`;
        if (!absolutePath.match(/^\/\w+\//)) {
            throw new Error('The path must begin with a bucket.');
        }
        return [app, absolutePath];
    }
    // Has local part?
    return [null, arg];
}
function isDirectory(db, path) {
    if (db) {
        return Promise.resolve(path.endsWith('/'));
    }
    return helper_1.isDir(path);
}
function extractFilename(path) {
    const index = path.lastIndexOf('/');
    if (index >= 0) {
        return path.substring(index + 1);
    }
    return path;
}
function removeTrailingSlash(path) {
    if (path.endsWith('/')) {
        return path.substring(0, path.length - 1);
    }
    return path;
}
function normalizeArgs(sourceDB, sourcePath, destDB, destPath) {
    return isDirectory(destDB, destPath).then((directory) => {
        let destination = destPath;
        if (directory) {
            const sourceFilename = extractFilename(sourcePath);
            destination = `${removeTrailingSlash(destPath)}/${sourceFilename}`;
        }
        return [sourcePath, destination];
    });
}
function login(sourceApp, destApp) {
    if (sourceApp) {
        if (destApp) {
            return Promise.all([account.login({ app: sourceApp }), account.login({ app: destApp })]);
        }
        return Promise.all([account.login({ app: sourceApp }), null]);
    }
    if (destApp) {
        return Promise.all([null, account.login({ app: destApp })]);
    }
    return Promise.resolve([null, null]);
}
function streamFrom(db, path) {
    if (db) {
        const file = new db.File({ path });
        return file.loadMetadata().then(() => Promise.all([file.download({ type: 'stream' }), file.size]));
    }
    return helper_1.stat(path).then((st) => {
        if (!st || !st.isFile()) {
            throw new Error(`${path} is not a valid file.`);
        }
        return [fs.createReadStream(path), st.size];
    });
}
function streamTo(db, path, rs, size) {
    if (db) {
        const file = new db.File({
            path, data: rs, size, type: 'stream',
        });
        return file.upload({ force: true });
    }
    return new Promise((resolve, reject) => {
        const ws = fs.createWriteStream(path);
        rs.on('end', resolve);
        rs.on('error', reject);
        rs.pipe(ws);
    });
}
/**
 * Copies from arbitrary location to each other.
 */
function copy(args) {
    // TODO: Split arguments with destructure in the future
    const [sourceApp, sourcePath] = splitArg(args.source);
    const [destApp, destPath] = splitArg(args.dest);
    return login(sourceApp, destApp)
        .then(([sourceDB, destDB]) => normalizeArgs(sourceDB, sourcePath, destDB, destPath)
        .then(([nSourcePath, nDestPath]) => streamFrom(sourceDB, nSourcePath)
        .then(([rs, size]) => streamTo(destDB, nDestPath, rs, size))));
}
exports.copy = copy;
