// baqend/cli/deploy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deploy = void 0;
/* eslint-disable no-await-in-loop, no-return-await, @typescript-eslint/return-await */
const fs_1 = __importDefault(require("fs"));
const glob_1 = __importDefault(require("glob"));
const path_1 = require("path");
const readline_1 = __importDefault(require("readline"));
const account = __importStar(require("./account"));
const schema_1 = require("./schema");
const helper_1 = require("./helper");
const handlerTypes = ['update', 'insert', 'delete', 'validate'];
function deploy(args) {
    return account.login(args).then((db) => {
        const promises = [];
        if ((!args.code && !args.files) || (args.code && args.files)) {
            promises.push(deployFiles(db, args.bucketPath, args.fileDir, args.fileGlob));
            promises.push(deployCode(db, args.codeDir));
        }
        else if (args.code) {
            promises.push(deployCode(db, args.codeDir));
        }
        else if (args.files) {
            promises.push(deployFiles(db, args.bucketPath, args.fileDir, args.fileGlob));
        }
        if (args.schema) {
            promises.push(schema_1.uploadSchema(db));
        }
        return Promise.all(promises);
    });
}
exports.deploy = deploy;
function deployFiles(db, path, cwd, pattern) {
    let normalizedPath = path;
    while (normalizedPath.length && normalizedPath.charAt(0) === '/')
        normalizedPath = normalizedPath.substring(1);
    while (normalizedPath.length && normalizedPath.charAt(normalizedPath.length - 1) === '/')
        normalizedPath = normalizedPath.substring(0, normalizedPath.length - 1);
    if (!normalizedPath.length) {
        console.error(`Invalid bucket-path ${normalizedPath}`);
        return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
        glob_1.default(pattern, { nodir: true, cwd }, (er, files) => {
            if (er)
                reject(er);
            else
                resolve(files);
        });
    })
        .then((files) => uploadFiles(db, normalizedPath, files, cwd))
        .then((result) => {
        if (result && result.length > 0) {
            console.log('File deployment completed.');
        }
        else {
            console.warn('Your specified upload folder is empty, no files were uploaded.');
        }
    });
}
function deployCode(db, codePath) {
    return helper_1.readDir(codePath)
        .catch(() => {
        throw new Error(`Your specified backend code folder ${codePath} is empty, no backend code was deployed.`);
    })
        .then((fileNames) => Promise.all(fileNames
        .map((fileName) => helper_1.stat(path_1.join(codePath, fileName))
        .then((st) => {
        if (st.isDirectory()) {
            return uploadHandler(db, fileName, codePath);
        }
        return uploadCode(db, fileName, codePath);
    })))
        .then(() => {
        console.log('Code deployment completed.');
    })
        .catch((e) => {
        throw new Error(`Failed to deploy code: ${e.message}`);
    }));
}
function uploadHandler(db, directoryName, codePath) {
    const bucket = directoryName;
    if (!db[bucket])
        return Promise.resolve();
    return helper_1.readDir(path_1.join(codePath, directoryName)).then((fileNames) => Promise.all(fileNames
        .filter((fileName) => !fileName.startsWith('.'))
        .map((fileName) => {
        const handlerType = fileName.replace(/\.(js|es\d*)$/, '');
        if (handlerTypes.indexOf(handlerType) === -1)
            return Promise.resolve();
        return helper_1.readFile(path_1.join(codePath, directoryName, fileName), 'utf-8')
            .then((file) => db.code.saveCode(bucket, handlerType, file))
            .then(() => console.log(`${handlerType} handler for ${bucket} deployed.`));
    })));
}
function uploadCode(db, name, codePath) {
    if (name.startsWith('.'))
        return Promise.resolve();
    const moduleName = name.replace(/\.(js|es\d*)$/, '');
    return helper_1.readFile(path_1.join(codePath, name), 'utf-8').then((file) => db.code.saveCode(moduleName, 'module', file)).then(() => {
        console.log(`Module ${moduleName} deployed.`);
    });
}
function uploadFiles(db, bucket, files, cwd) {
    const isTty = process.stdout.isTTY;
    let index = 0;
    const uploads = [];
    for (let i = 0; i < 10 && i < files.length; i += 1) {
        uploads.push(upload());
    }
    if (!isTty) {
        console.log(`Uploading ${files.length} files.`);
    }
    return Promise.all(uploads);
    function upload() {
        return __awaiter(this, void 0, void 0, function* () {
            while (index < files.length) {
                if (isTty) {
                    if (index > 0) {
                        readline_1.default.clearLine(process.stdout, 0);
                        readline_1.default.cursorTo(process.stdout, 0);
                    }
                    process.stdout.write(`Uploading file ${(index + 1)} of ${files.length}`);
                }
                const file = files[index];
                index += 1;
                if (isTty && index === files.length) {
                    console.log(''); // add a final linebreak
                    return;
                }
                yield uploadFile(db, bucket, file, cwd);
            }
        });
    }
}
function uploadFile(db, bucket, filePath, cwd) {
    return __awaiter(this, void 0, void 0, function* () {
        const fullFilePath = path_1.join(cwd, filePath);
        const st = yield helper_1.stat(fullFilePath);
        if (!st || st.isDirectory()) {
            return null;
        }
        for (let retires = 3;; retires -= 1) {
            try {
                const file = new db.File({
                    path: `/${bucket}/${filePath}`, data: fs_1.default.createReadStream(fullFilePath), size: st.size, type: 'stream',
                });
                return yield file.upload({ force: true });
            }
            catch (e) {
                if (retires <= 0) {
                    console.warn(`Failed to upload file ${filePath}. ${retires} retries left.`);
                    throw new Error(`Failed to upload file ${filePath}: ${e.message}`);
                }
            }
        }
    });
}
