// baqend/cli/helper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNativeClassNamespace = exports.readInput = exports.ensureDir = exports.isFile = exports.isDir = exports.stat = exports.nativeNamespaces = exports.mkdir = exports.readFile = exports.writeFile = exports.readDir = void 0;
const readline_1 = __importDefault(require("readline"));
const stream_1 = require("stream");
const util_1 = require("util");
const fs = __importStar(require("fs"));
exports.readDir = util_1.promisify(fs.readdir);
exports.writeFile = util_1.promisify(fs.writeFile);
exports.readFile = util_1.promisify(fs.readFile);
exports.mkdir = util_1.promisify(fs.mkdir);
exports.nativeNamespaces = ['logs', 'speedKit', 'rum', 'jobs'];
/**
 * Returns the stats for the given path
 * @param path
 */
function stat(path) {
    return new Promise((resolve, reject) => {
        fs.stat(path, (err, st) => {
            if (!err) {
                resolve(st);
            }
            else if (err.code === 'ENOENT') {
                resolve(null);
            }
            else {
                reject(err);
            }
        });
    });
}
exports.stat = stat;
/**
 * Indicates if the given path is a directory
 * @param path
 */
function isDir(path) {
    return stat(path).then((s) => !!s && s.isDirectory());
}
exports.isDir = isDir;
/**
 * Indicates if the given path is a file
 * @param path
 */
function isFile(path) {
    return stat(path).then((s) => !!s && s.isFile());
}
exports.isFile = isFile;
/**
 * Creates a direcotry or ensures that it exists.
 *
 * @param {string} dir The path where a directory should exist.
 * @return {Promise<void>} Resolves when the given directory is existing.
 */
function ensureDir(dir) {
    return isDir(dir).then((directory) => {
        if (!directory) {
            return exports.mkdir(dir, { recursive: true });
        }
        return Promise.resolve();
    });
}
exports.ensureDir = ensureDir;
function readInput(question, hidden = false) {
    let muted = false;
    const mutableStdout = new stream_1.Writable({
        write(chunk, encoding, callback) {
            if (!muted) {
                process.stdout.write(chunk, encoding);
            }
            callback();
        },
    });
    const rl = readline_1.default.createInterface({
        input: process.stdin,
        output: mutableStdout,
        terminal: true,
    });
    return new Promise((resolve, reject) => {
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer);
        });
        rl.on('SIGINT', () => {
            rl.close();
            reject(new Error('Input is aborted.'));
        });
        muted = hidden;
    });
}
exports.readInput = readInput;
function isNativeClassNamespace(className) {
    const [namespace] = className.split('.');
    return exports.nativeNamespaces.includes(namespace);
}
exports.isNativeClassNamespace = isNativeClassNamespace;
